// @salesforce/plugin-source/lib/stash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stash = void 0;
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const fs = require("fs");
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
class Stash {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    constructor() { }
    /**
     * Returns the value for the given stash key.
     *
     * @param key The StashKey
     * @returns The stash value for the given key.
     */
    static get(key) {
        const stash = Stash.getStashFile();
        Stash.logger.debug(`Getting ${key} data from ${stash.getPath()}`);
        return stash.get(key);
    }
    /**
     * Returns the `StashKey` used by the Command.
     *
     * Within a command class you would typically call this as:
     *
     * `const stashEntry = Stash.get(Stash.getKey(this.id));`
     *
     * @param commandId The oclif Command.id.  E.g., `this.id`
     * @returns the `StashKey` to use for `Stash.get()` and `Stash.set()`
     */
    static getKey(commandId) {
        const key = Stash.keyMap[commandId];
        if (!key) {
            const messages = core_1.Messages.load('@salesforce/plugin-source', 'stash', ['InvalidStashKey']);
            throw new core_1.SfError(messages.getMessage('InvalidStashKey', [commandId]), 'InvalidStashKey');
        }
        return key;
    }
    /**
     * Sets deploy/retrieve data in the stash.
     *
     * @param key the `StashKey` for setting stashed deploy/retrieve values
     * @param data the `StashData` to persist.
     */
    static set(key, data) {
        const stash = Stash.getStashFile();
        stash.set(key, data);
        Stash.logger.debug(`Stashing ${key} data: ${JSON.stringify(data)} in ${stash.getPath()}`);
        stash.writeSync();
    }
    /**
     * Clears all stash file entries.
     */
    static clear() {
        const stash = Stash.getStashFile();
        Stash.logger.debug(`Clearing all stash contents in: ${stash.getPath()}`);
        stash.clear();
        stash.writeSync();
    }
    static init() {
        Stash.logger = core_1.Logger.childFromRoot('source-plugin-stash');
        Stash.instance = new core_1.ConfigFile({ isGlobal: true, filename: 'stash.json' });
    }
    static getStashFile() {
        var _a;
        try {
            if (!Stash.instance) {
                Stash.init();
            }
            if (Stash.instance.existsSync()) {
                // Always read from the stash file before returning it
                Stash.instance.readSync(true);
            }
            else {
                Stash.instance.writeSync();
            }
        }
        catch (err) {
            const error = err;
            if (error.name === 'JsonParseError') {
                const stashFilePath = (_a = Stash.instance) === null || _a === void 0 ? void 0 : _a.getPath();
                const corruptFilePath = `${stashFilePath}_corrupted_${Date.now()}`;
                fs.renameSync(stashFilePath, corruptFilePath);
                const messages = core_1.Messages.load('@salesforce/plugin-source', 'stash', ['InvalidStashKey']);
                throw new core_1.SfError(`${messages.getMessage('InvalidStashKey', [corruptFilePath])}\n\n${error.message}`, 'InvalidStashFile', [], error);
            }
            throw core_1.SfError.wrap(error);
        }
        return Stash.instance;
    }
}
exports.Stash = Stash;
Stash.KEYS = {
    MDAPI_DEPLOY: 'MDAPI_DEPLOY',
    MDAPI_RETRIEVE: 'MDAPI_RETRIEVE',
    SOURCE_DEPLOY: 'SOURCE_DEPLOY',
};
// A map of Command.id to stash keys
Stash.keyMap = {
    'force:mdapi:deploy': Stash.KEYS.MDAPI_DEPLOY,
    'force:mdapi:deploy:cancel': Stash.KEYS.MDAPI_DEPLOY,
    'force:mdapi:deploy:report': Stash.KEYS.MDAPI_DEPLOY,
    'force:mdapi:beta:deploy': Stash.KEYS.MDAPI_DEPLOY,
    'force:mdapi:beta:deploy:cancel': Stash.KEYS.MDAPI_DEPLOY,
    'force:mdapi:beta:deploy:report': Stash.KEYS.MDAPI_DEPLOY,
    'force:source:deploy': Stash.KEYS.SOURCE_DEPLOY,
    'force:source:deploy:cancel': Stash.KEYS.SOURCE_DEPLOY,
    'force:source:deploy:report': Stash.KEYS.SOURCE_DEPLOY,
    'force:source:delete': Stash.KEYS.SOURCE_DEPLOY,
    'force:mdapi:retrieve': Stash.KEYS.MDAPI_RETRIEVE,
    'force:mdapi:retrieve:report': Stash.KEYS.MDAPI_RETRIEVE,
    'force:mdapi:beta:retrieve': Stash.KEYS.MDAPI_RETRIEVE,
    'force:mdapi:beta:retrieve:report': Stash.KEYS.MDAPI_RETRIEVE,
};
