// jsforce/browser/api/streaming.js
import _getIterator from "@babel/runtime-corejs3/core-js/get-iterator";
import _getIteratorMethod from "@babel/runtime-corejs3/core-js/get-iterator-method";
import _Symbol from "@babel/runtime-corejs3/core-js-stable/symbol";
import _Array$from from "@babel/runtime-corejs3/core-js-stable/array/from";
import _sliceInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/slice";
import _Reflect$construct from "@babel/runtime-corejs3/core-js-stable/reflect/construct";
import "core-js/modules/es.array.join";
import "core-js/modules/es.function.name";
import _indexOfInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/index-of";
import _assertThisInitialized from "@babel/runtime-corejs3/helpers/assertThisInitialized";
import _inherits from "@babel/runtime-corejs3/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime-corejs3/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime-corejs3/helpers/getPrototypeOf";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
import _Array$isArray from "@babel/runtime-corejs3/core-js-stable/array/is-array";
import "regenerator-runtime/runtime";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime-corejs3/helpers/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/createClass";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof _Symbol === "undefined" || _getIteratorMethod(o) == null) { if (_Array$isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = _getIterator(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { var _context3; if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = _sliceInstanceProperty(_context3 = Object.prototype.toString.call(o)).call(_context3, 8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return _Array$from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = _Reflect$construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !_Reflect$construct) return false; if (_Reflect$construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(_Reflect$construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * @file Manages Streaming APIs
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */
import { EventEmitter } from 'events';
import { Client, Subscription } from 'faye';
import { registerModule } from '../jsforce';
import * as StreamingExtension from './streaming/extension';
/**
 *
 */

export { Client, Subscription };
/*--------------------------------------------*/

/**
 * Streaming API topic class
 */

var Topic = /*#__PURE__*/function () {
  function Topic(streaming, name) {
    _classCallCheck(this, Topic);

    _defineProperty(this, "_streaming", void 0);

    _defineProperty(this, "name", void 0);

    this._streaming = streaming;
    this.name = name;
  }
  /**
   * Subscribe listener to topic
   */


  _createClass(Topic, [{
    key: "subscribe",
    value: function subscribe(listener) {
      return this._streaming.subscribe(this.name, listener);
    }
    /**
     * Unsubscribe listener from topic
     */

  }, {
    key: "unsubscribe",
    value: function unsubscribe(subscr) {
      this._streaming.unsubscribe(this.name, subscr);

      return this;
    }
  }]);

  return Topic;
}();
/*--------------------------------------------*/

/**
 * Streaming API Generic Streaming Channel
 */


var Channel = /*#__PURE__*/function () {
  function Channel(streaming, name) {
    _classCallCheck(this, Channel);

    _defineProperty(this, "_streaming", void 0);

    _defineProperty(this, "_id", void 0);

    _defineProperty(this, "name", void 0);

    this._streaming = streaming;
    this.name = name;
  }
  /**
   * Subscribe to channel
   */


  _createClass(Channel, [{
    key: "subscribe",
    value: function subscribe(listener) {
      return this._streaming.subscribe(this.name, listener);
    }
  }, {
    key: "unsubscribe",
    value: function unsubscribe(subscr) {
      this._streaming.unsubscribe(this.name, subscr);

      return this;
    }
  }, {
    key: "push",
    value: function () {
      var _push = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(events) {
        var isArray, pushEvents, conn, id, channelUrl, rets;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                isArray = _Array$isArray(events);
                pushEvents = _Array$isArray(events) ? events : [events];
                conn = this._streaming._conn;

                if (!this._id) {
                  this._id = conn.sobject('StreamingChannel').findOne({
                    Name: this.name
                  }, ['Id']).then(function (rec) {
                    return rec === null || rec === void 0 ? void 0 : rec.Id;
                  });
                }

                _context.next = 6;
                return this._id;

              case 6:
                id = _context.sent;

                if (id) {
                  _context.next = 9;
                  break;
                }

                throw new Error("No streaming channel available for name: ".concat(this.name));

              case 9:
                channelUrl = "/sobjects/StreamingChannel/".concat(id, "/push");
                _context.next = 12;
                return conn.requestPost(channelUrl, {
                  pushEvents: pushEvents
                });

              case 12:
                rets = _context.sent;
                return _context.abrupt("return", isArray ? rets : rets[0]);

              case 14:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function push(_x) {
        return _push.apply(this, arguments);
      }

      return push;
    }()
  }]);

  return Channel;
}();
/*--------------------------------------------*/

/**
 * Streaming API class
 */


export var Streaming = /*#__PURE__*/function (_EventEmitter) {
  _inherits(Streaming, _EventEmitter);

  var _super = _createSuper(Streaming);

  /**
   *
   */
  function Streaming(conn) {
    var _this;

    _classCallCheck(this, Streaming);

    _this = _super.call(this);

    _defineProperty(_assertThisInitialized(_this), "_conn", void 0);

    _defineProperty(_assertThisInitialized(_this), "_topics", {});

    _defineProperty(_assertThisInitialized(_this), "_fayeClients", {});

    _this._conn = conn;
    return _this;
  }
  /* @private */


  _createClass(Streaming, [{
    key: "_createClient",
    value: function _createClient(forChannelName, extensions) {
      var _context2;

      // forChannelName is advisory, for an API workaround. It does not restrict or select the channel.
      var needsReplayFix = typeof forChannelName === 'string' && _indexOfInstanceProperty(forChannelName).call(forChannelName, '/u/') === 0;
      var endpointUrl = [this._conn.instanceUrl, // special endpoint "/cometd/replay/xx.x" is only available in 36.0.
      // See https://releasenotes.docs.salesforce.com/en-us/summer16/release-notes/rn_api_streaming_classic_replay.htm
      'cometd' + (needsReplayFix === true && this._conn.version === '36.0' ? '/replay' : ''), this._conn.version].join('/');
      var fayeClient = new Client(endpointUrl, {});
      fayeClient.setHeader('Authorization', 'OAuth ' + this._conn.accessToken);

      if (_Array$isArray(extensions)) {
        var _iterator = _createForOfIteratorHelper(extensions),
            _step;

        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var extension = _step.value;
            fayeClient.addExtension(extension);
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
      } // prevent streaming API server error


      var dispatcher = fayeClient._dispatcher;

      if (_indexOfInstanceProperty(_context2 = dispatcher.getConnectionTypes()).call(_context2, 'callback-polling') === -1) {
        dispatcher.selectTransport('long-polling');
        dispatcher._transport.batching = false;
      }

      return fayeClient;
    }
    /** @private **/

  }, {
    key: "_getFayeClient",
    value: function _getFayeClient(channelName) {
      var isGeneric = _indexOfInstanceProperty(channelName).call(channelName, '/u/') === 0;
      var clientType = isGeneric ? 'generic' : 'pushTopic';

      if (!this._fayeClients[clientType]) {
        this._fayeClients[clientType] = this._createClient(channelName);
      }

      return this._fayeClients[clientType];
    }
    /**
     * Get named topic
     */

  }, {
    key: "topic",
    value: function topic(name) {
      this._topics = this._topics || {};
      var topic = this._topics[name] = this._topics[name] || new Topic(this, name);
      return topic;
    }
    /**
     * Get channel for channel name
     */

  }, {
    key: "channel",
    value: function channel(name) {
      return new Channel(this, name);
    }
    /**
     * Subscribe topic/channel
     */

  }, {
    key: "subscribe",
    value: function subscribe(name, listener) {
      var channelName = _indexOfInstanceProperty(name).call(name, '/') === 0 ? name : '/topic/' + name;

      var fayeClient = this._getFayeClient(channelName);

      return fayeClient.subscribe(channelName, listener);
    }
    /**
     * Unsubscribe topic
     */

  }, {
    key: "unsubscribe",
    value: function unsubscribe(name, subscription) {
      var channelName = _indexOfInstanceProperty(name).call(name, '/') === 0 ? name : '/topic/' + name;

      var fayeClient = this._getFayeClient(channelName);

      fayeClient.unsubscribe(channelName, subscription);
      return this;
    }
    /**
     * Create a Streaming client, optionally with extensions
     *
     * See Faye docs for implementation details: https://faye.jcoglan.com/browser/extensions.html
     *
     * Example usage:
     *
     * ```javascript
     * const jsforce = require('jsforce');
     *
     * // Establish a Salesforce connection. (Details elided)
     * const conn = new jsforce.Connection({ … });
     *
     * const fayeClient = conn.streaming.createClient();
     *
     * const subscription = fayeClient.subscribe(channel, data => {
     *   console.log('topic received data', data);
     * });
     *
     * subscription.cancel();
     * ```
     *
     * Example with extensions, using Replay & Auth Failure extensions in a server-side Node.js app:
     *
     * ```javascript
     * const jsforce = require('jsforce');
     * const { StreamingExtension } = require('jsforce/api/streaming');
     *
     * // Establish a Salesforce connection. (Details elided)
     * const conn = new jsforce.Connection({ … });
     *
     * const channel = "/event/My_Event__e";
     * const replayId = -2; // -2 is all retained events
     *
     * const exitCallback = () => process.exit(1);
     * const authFailureExt = new StreamingExtension.AuthFailure(exitCallback);
     *
     * const replayExt = new StreamingExtension.Replay(channel, replayId);
     *
     * const fayeClient = conn.streaming.createClient([
     *   authFailureExt,
     *   replayExt
     * ]);
     *
     * const subscription = fayeClient.subscribe(channel, data => {
     *   console.log('topic received data', data);
     * });
     *
     * subscription.cancel();
     * ```
     */

  }, {
    key: "createClient",
    value: function createClient(extensions) {
      return this._createClient(null, extensions);
    }
  }]);

  return Streaming;
}(EventEmitter);
export { StreamingExtension };
/*--------------------------------------------*/

/*
 * Register hook in connection instantiation for dynamically adding this API module features
 */

registerModule('streaming', function (conn) {
  return new Streaming(conn);
});
export default Streaming;
