// @salesforce/apex-node/lib/src/tests/types.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApexTestQueueItemStatus = exports.ApexTestRunResultStatus = exports.ApexTestResultOutcome = exports.ResultFormat = exports.TestLevel = void 0;
var TestLevel;
(function (TestLevel) {
    /**
     * All tests in your org are run, except the ones that originate from installed managed packages
     */
    TestLevel["RunLocalTests"] = "RunLocalTests";
    /**
     * All tests are in your org and in installed managed packages are run
     */
    TestLevel["RunAllTestsInOrg"] = "RunAllTestsInOrg";
    /**
     * Only the tests that you specify are run
     */
    TestLevel["RunSpecifiedTests"] = "RunSpecifiedTests";
})(TestLevel = exports.TestLevel || (exports.TestLevel = {}));
var ResultFormat;
(function (ResultFormat) {
    ResultFormat["junit"] = "junit";
    ResultFormat["tap"] = "tap";
    ResultFormat["json"] = "json";
    ResultFormat["human"] = "human";
})(ResultFormat = exports.ResultFormat || (exports.ResultFormat = {}));
var ApexTestResultOutcome;
(function (ApexTestResultOutcome) {
    ApexTestResultOutcome["Pass"] = "Pass";
    ApexTestResultOutcome["Fail"] = "Fail";
    ApexTestResultOutcome["CompileFail"] = "CompileFail";
    ApexTestResultOutcome["Skip"] = "Skip";
})(ApexTestResultOutcome = exports.ApexTestResultOutcome || (exports.ApexTestResultOutcome = {}));
var ApexTestRunResultStatus;
(function (ApexTestRunResultStatus) {
    ApexTestRunResultStatus["Queued"] = "Queued";
    ApexTestRunResultStatus["Processing"] = "Processing";
    ApexTestRunResultStatus["Aborted"] = "Aborted";
    ApexTestRunResultStatus["Passed"] = "Passed";
    ApexTestRunResultStatus["Failed"] = "Failed";
    ApexTestRunResultStatus["Completed"] = "Completed";
    ApexTestRunResultStatus["Skipped"] = "Skipped";
})(ApexTestRunResultStatus = exports.ApexTestRunResultStatus || (exports.ApexTestRunResultStatus = {}));
var ApexTestQueueItemStatus;
(function (ApexTestQueueItemStatus) {
    ApexTestQueueItemStatus["Holding"] = "Holding";
    ApexTestQueueItemStatus["Queued"] = "Queued";
    ApexTestQueueItemStatus["Preparing"] = "Preparing";
    ApexTestQueueItemStatus["Processing"] = "Processing";
    ApexTestQueueItemStatus["Aborted"] = "Aborted";
    ApexTestQueueItemStatus["Completed"] = "Completed";
    ApexTestQueueItemStatus["Failed"] = "Failed";
})(ApexTestQueueItemStatus = exports.ApexTestQueueItemStatus || (exports.ApexTestQueueItemStatus = {}));
