// jsforce/lib/util/function.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.isObject = isObject;
exports.isMapObject = isMapObject;
exports.isFunction = isFunction;
exports.isNumber = isNumber;
exports.isPromiseLike = isPromiseLike;
exports.identityFunc = identityFunc;
exports.emptyFunc = emptyFunc;

/**
 *
 */
function isObject(v) {
  const t = typeof v;
  return v != null && (t == 'object' || t == 'function');
}
/**
 *
 */


function isMapObject(v) {
  const t = typeof v;
  return v != null && t == 'object';
}
/**
 *
 */


function isFunction(v) {
  return typeof v == 'function';
}
/**
 *
 */


function isNumber(v) {
  return typeof v == 'number';
}
/**
 * Detect whether the value has CommonJS Promise/A+ interface or not
 */


function isPromiseLike(v) {
  return isObject(v) && isFunction(v.then);
}
/**
 *
 */


function identityFunc(a) {
  return a;
}
/**
 *
 */


function emptyFunc() {}
