// @salesforce/plugin-templates/lib/commands/force/apex/class/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const apexClassGenerator_1 = require("@salesforce/templates/lib/generators/apexClassGenerator");
const utils_1 = require("@salesforce/templates/lib/utils");
const utils_2 = require("../../../../utils");
const apexClassFileSuffix = /.cls$/;
class ApexClass extends utils_2.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.runGenerator(apexClassGenerator_1.default);
        });
    }
}
exports.default = ApexClass;
ApexClass.description = utils_2.MessageUtil.buildDescription('ApexClassDescription', false);
ApexClass.examples = [
    '$ sfdx force:apex:class:create -n MyClass',
    '$ sfdx force:apex:class:create -n MyClass -d classes'
];
ApexClass.help = utils_2.MessageUtil.buildHelpText(ApexClass.examples, false);
ApexClass.longDescription = utils_2.MessageUtil.get('ApexClassLongDescription');
ApexClass.flagsConfig = {
    classname: command_1.flags.string({
        char: 'n',
        description: utils_2.MessageUtil.get('ApexClassNameFlagDescription'),
        longDescription: utils_2.MessageUtil.get('ApexClassNameFlagLongDescription'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: utils_2.MessageUtil.get('TemplateFlagDescription'),
        longDescription: utils_2.MessageUtil.get('TemplateFlagLongDescription'),
        default: 'DefaultApexClass',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(apexClassFileSuffix, 'apexclass')
    }),
    outputdir: command_1.flags.string({
        char: 'd',
        description: utils_2.MessageUtil.get('OutputDirFlagDescription'),
        longDescription: utils_2.MessageUtil.get('OutputDirFlagLongDescription'),
        default: '.'
    }),
    apiversion: command_1.flags.builtin()
};
