// @salesforce/plugin-trust/lib/hooks/verifyInstallSignature.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.hook = exports.VerificationConfigBuilder = void 0;
const core_1 = require("@salesforce/core");
const cli_ux_1 = require("cli-ux");
const installationVerification_1 = require("../shared/installationVerification");
const NpmName_1 = require("../shared/NpmName");
/**
 * Build a VerificationConfig. Useful for testing.
 */
class VerificationConfigBuilder {
    static build(npmName, configContext) {
        const vConfig = new installationVerification_1.VerificationConfig();
        vConfig.verifier = new installationVerification_1.InstallationVerification().setPluginNpmName(npmName).setConfig(configContext);
        vConfig.log = cli_ux_1.cli.log.bind(cli_ux_1.cli);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        vConfig.prompt = cli_ux_1.cli.prompt.bind(cli_ux_1.cli);
        return vConfig;
    }
    static buildForRepo() {
        const vConfig = new installationVerification_1.VerificationConfig();
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        vConfig.prompt = cli_ux_1.cli.prompt.bind(cli_ux_1.cli);
        return vConfig;
    }
}
exports.VerificationConfigBuilder = VerificationConfigBuilder;
const hook = async function (options) {
    if (options.plugin && options.plugin.type === 'npm') {
        const logger = await core_1.Logger.child('verifyInstallSignature');
        const plugin = options.plugin;
        logger.debug('parsing npm name');
        const npmName = NpmName_1.NpmName.parse(plugin.name);
        logger.debug(`npmName components: ${JSON.stringify(npmName, null, 4)}`);
        npmName.tag = plugin.tag || 'latest';
        if (/^v[0-9].*/.test(npmName.tag)) {
            npmName.tag = npmName.tag.slice(1);
        }
        const configContext = {
            cacheDir: options.config.cacheDir,
            configDir: options.config.configDir,
            dataDir: options.config.dataDir,
            cliRoot: options.config.root,
        };
        const vConfig = VerificationConfigBuilder.build(npmName, configContext);
        logger.debug('finished building the VerificationConfigBuilder');
        try {
            logger.debug('doing verification');
            await (0, installationVerification_1.doInstallationCodeSigningVerification)(configContext, { plugin: plugin.name, tag: plugin.tag }, vConfig);
            cli_ux_1.cli.log('Finished digital signature check.');
        }
        catch (error) {
            const err = error;
            logger.debug(err.message);
            this.error(err);
        }
    }
    else {
        await (0, installationVerification_1.doPrompt)(VerificationConfigBuilder.buildForRepo());
    }
};
exports.hook = hook;
exports.default = exports.hook;
