// salesforce-alm/dist/commands/force/package/version/delete.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersionDeleteCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../../lib/core/constants");
class PackageVersionDeleteCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        context.flagsConfig = this.statics.flagsConfig;
        const PackageVersionDeleteCommandImpl = require('../../../../lib/package/packageVersionDeleteCommand');
        const heroku = require('heroku-cli-util');
        return this.execLegacyCommand(new PackageVersionDeleteCommandImpl(heroku.prompt), context);
    }
}
exports.PackageVersionDeleteCommand = PackageVersionDeleteCommand;
PackageVersionDeleteCommand.theDescription = messages.getMessage('cliDescription', [], 'package_version_delete');
PackageVersionDeleteCommand.longDescription = messages.getMessage('cliLongDescription', [], 'package_version_delete');
PackageVersionDeleteCommand.help = messages.getMessage('help', [], 'package_version_delete');
PackageVersionDeleteCommand.orgType = consts.DEFAULT_DEV_HUB_USERNAME;
PackageVersionDeleteCommand.requiresDevhubUsername = true;
PackageVersionDeleteCommand.requiresProject = true;
PackageVersionDeleteCommand.flagsConfig = {
    noprompt: command_1.flags.boolean({
        char: 'n',
        description: messages.getMessage('noPrompt', [], 'package_version_delete'),
        longDescription: messages.getMessage('noPrompt', [], 'package_version_delete'),
        required: false,
    }),
    package: command_1.flags.string({
        char: 'p',
        description: messages.getMessage('package', [], 'package_version_delete'),
        longDescription: messages.getMessage('packageLong', [], 'package_version_delete'),
        required: true,
    }),
    undelete: command_1.flags.boolean({
        description: messages.getMessage('undelete', [], 'package_version_delete'),
        longDescription: messages.getMessage('undeleteLong', [], 'package_version_delete'),
        required: false,
        hidden: true,
    }),
};
