// salesforce-alm/dist/commands/force/package/version/report.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersionReportCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../../lib/core/constants");
class PackageVersionReportCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const PackageVersionReportCommandImpl = require('../../../../lib/package/packageVersionReportCommand');
        return this.execLegacyCommand(new PackageVersionReportCommandImpl(), context);
    }
}
exports.PackageVersionReportCommand = PackageVersionReportCommand;
PackageVersionReportCommand.theDescription = messages.getMessage('cliDescription', [], 'package_version_report');
PackageVersionReportCommand.longDescription = messages.getMessage('cliLongDescription', [], 'package_version_report');
PackageVersionReportCommand.help = messages.getMessage('help', [], 'package_version_report');
PackageVersionReportCommand.orgType = consts.DEFAULT_DEV_HUB_USERNAME;
PackageVersionReportCommand.requiresDevhubUsername = true;
PackageVersionReportCommand.requiresProject = true;
PackageVersionReportCommand.flagsConfig = {
    package: command_1.flags.string({
        char: 'p',
        description: messages.getMessage('package', [], 'package_version_report'),
        longDescription: messages.getMessage('packageLong', [], 'package_version_report'),
        required: true,
    }),
    verbose: command_1.flags.builtin({
        description: messages.getMessage('verboseDescription', [], 'package_version_report'),
        longDescription: messages.getMessage('verboseLongDescription', [], 'package_version_report'),
    }),
};
