// salesforce-alm/dist/commands/force/package/version/update.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageVersionUpdateCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../../lib/core/constants");
class PackageVersionUpdateCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        context.flagsConfig = this.statics.flagsConfig;
        const PackageVersionUpdateCommandImpl = require('../../../../lib/package/packageVersionUpdateCommand');
        const packageVersionUpdateCommand = new PackageVersionUpdateCommandImpl();
        return this.execLegacyCommand(packageVersionUpdateCommand, context);
    }
}
exports.PackageVersionUpdateCommand = PackageVersionUpdateCommand;
PackageVersionUpdateCommand.theDescription = messages.getMessage('cliDescription', [], 'package_version_update');
PackageVersionUpdateCommand.longDescription = messages.getMessage('cliLongDescription', [], 'package_version_update');
PackageVersionUpdateCommand.help = messages.getMessage('help', [], 'package_version_update');
PackageVersionUpdateCommand.orgType = consts.DEFAULT_DEV_HUB_USERNAME;
PackageVersionUpdateCommand.requiresDevhubUsername = true;
PackageVersionUpdateCommand.requiresProject = true;
PackageVersionUpdateCommand.flagsConfig = {
    package: command_1.flags.string({
        char: 'p',
        description: messages.getMessage('package', [], 'package_version_update'),
        longDescription: messages.getMessage('packageLong', [], 'package_version_update'),
        required: true,
    }),
    versionname: command_1.flags.string({
        char: 'a',
        description: messages.getMessage('name', [], 'package_version_update'),
        longDescription: messages.getMessage('nameLong', [], 'package_version_update'),
        required: false,
    }),
    versiondescription: command_1.flags.string({
        char: 'e',
        description: messages.getMessage('description', [], 'package_version_update'),
        longDescription: messages.getMessage('descriptionLong', [], 'package_version_update'),
        required: false,
    }),
    branch: command_1.flags.string({
        char: 'b',
        description: messages.getMessage('branch', [], 'package_version_update'),
        longDescription: messages.getMessage('branchLong', [], 'package_version_update'),
        required: false,
    }),
    tag: command_1.flags.string({
        char: 't',
        description: messages.getMessage('tag', [], 'package_version_update'),
        longDescription: messages.getMessage('tagLong', [], 'package_version_update'),
        required: false,
    }),
    installationkey: command_1.flags.string({
        char: 'k',
        description: messages.getMessage('key', [], 'package_version_update'),
        longDescription: messages.getMessage('longKey', [], 'package_version_update'),
        required: false,
    }),
};
