// jsforce/browser/quick-action.js
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
import "regenerator-runtime/runtime";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime-corejs3/helpers/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/createClass";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";

/**
 * @file Represents Salesforce QuickAction
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */

/**
 * type definitions
 */

/**
 * A class for quick action
 */
export var QuickAction = /*#__PURE__*/function () {
  /**
   *
   */
  function QuickAction(conn, path) {
    _classCallCheck(this, QuickAction);

    _defineProperty(this, "_conn", void 0);

    _defineProperty(this, "_path", void 0);

    this._conn = conn;
    this._path = path;
  }
  /**
   * Describe the action's information (including layout, etc.)
   */


  _createClass(QuickAction, [{
    key: "describe",
    value: function () {
      var _describe = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        var url, body;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                url = "".concat(this._path, "/describe");
                _context.next = 3;
                return this._conn.request(url);

              case 3:
                body = _context.sent;
                return _context.abrupt("return", body);

              case 5:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function describe() {
        return _describe.apply(this, arguments);
      }

      return describe;
    }()
    /**
     * Retrieve default field values in the action (for given record, if specified)
     */

  }, {
    key: "defaultValues",
    value: function () {
      var _defaultValues = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(contextId) {
        var url, body;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                url = "".concat(this._path, "/defaultValues");

                if (contextId) {
                  url += "/".concat(contextId);
                }

                _context2.next = 4;
                return this._conn.request(url);

              case 4:
                body = _context2.sent;
                return _context2.abrupt("return", body);

              case 6:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function defaultValues(_x) {
        return _defaultValues.apply(this, arguments);
      }

      return defaultValues;
    }()
    /**
     * Execute the action for given context Id and record information
     */

  }, {
    key: "execute",
    value: function () {
      var _execute = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee3(contextId, record) {
        var requestBody, resBody;
        return _regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                requestBody = {
                  contextId: contextId,
                  record: record
                };
                _context3.next = 3;
                return this._conn.requestPost(this._path, requestBody);

              case 3:
                resBody = _context3.sent;
                return _context3.abrupt("return", resBody);

              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function execute(_x2, _x3) {
        return _execute.apply(this, arguments);
      }

      return execute;
    }()
  }]);

  return QuickAction;
}();
export default QuickAction;
