// jsforce/module/browser/jsonp.js
import "core-js/modules/es.promise";
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _setTimeout from "@babel/runtime-corejs3/core-js-stable/set-timeout";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import _indexOfInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/index-of";

/**
 *
 */
import { Transform } from 'stream';
let _index = 0;

async function processJsonpRequest(params, jsonpParam, timeout) {
  if (params.method.toUpperCase() !== 'GET') {
    throw new Error('JSONP only supports GET request.');
  }

  _index += 1;
  const cbFuncName = `_jsforce_jsonpCallback_${_index}`;
  const callbacks = window;
  let url = params.url;
  url += _indexOfInstanceProperty(url).call(url, '?') > 0 ? '&' : '?';
  url += `${jsonpParam}=${cbFuncName}`;
  const script = document.createElement('script');
  script.type = 'text/javascript';
  script.src = url;

  if (document.documentElement) {
    document.documentElement.appendChild(script);
  }

  let pid;

  try {
    const res = await new _Promise((resolve, reject) => {
      pid = _setTimeout(() => {
        reject(new Error('JSONP call time out.'));
      }, timeout);
      callbacks[cbFuncName] = resolve;
    });
    return {
      statusCode: 200,
      headers: {
        'content-type': 'application/json'
      },
      body: _JSON$stringify(res)
    };
  } finally {
    clearTimeout(pid);

    if (document.documentElement) {
      document.documentElement.removeChild(script);
    }

    delete callbacks[cbFuncName];
  }
}

function createRequest(jsonpParam = 'callback', timeout = 10000) {
  return params => {
    const stream = new Transform({
      transform(chunk, encoding, callback) {
        callback();
      },

      flush() {
        (async () => {
          const response = await processJsonpRequest(params, jsonpParam, timeout);
          stream.emit('response', response);
          stream.emit('complete', response);
          stream.push(response.body);
          stream.push(null);
        })();
      }

    });
    stream.end();
    return stream;
  };
}

export default {
  supported: typeof window !== 'undefined' && typeof document !== 'undefined',
  createRequest
};
