// @salesforce/apex-node/lib/src/reporters/index.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultWatermarks = exports.DefaultReportOptions = exports.CoverageReporter = exports.HumanReporter = exports.JUnitReporter = exports.TapReporter = void 0;
var tapReporter_1 = require("./tapReporter");
Object.defineProperty(exports, "TapReporter", { enumerable: true, get: function () { return tapReporter_1.TapReporter; } });
var junitReporter_1 = require("./junitReporter");
Object.defineProperty(exports, "JUnitReporter", { enumerable: true, get: function () { return junitReporter_1.JUnitReporter; } });
var humanReporter_1 = require("./humanReporter");
Object.defineProperty(exports, "HumanReporter", { enumerable: true, get: function () { return humanReporter_1.HumanReporter; } });
var coverageReporter_1 = require("./coverageReporter");
Object.defineProperty(exports, "CoverageReporter", { enumerable: true, get: function () { return coverageReporter_1.CoverageReporter; } });
Object.defineProperty(exports, "DefaultReportOptions", { enumerable: true, get: function () { return coverageReporter_1.DefaultReportOptions; } });
Object.defineProperty(exports, "DefaultWatermarks", { enumerable: true, get: function () { return coverageReporter_1.DefaultWatermarks; } });
