// @salesforce/apex-node/lib/src/reporters/tapReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TapReporter = void 0;
class TapReporter {
    format(testResult, epilog) {
        const results = this.buildTapResults(testResult);
        const testPointCount = results.length;
        let out = '';
        out = out.concat(`1..${testPointCount}\n`);
        results.forEach(testPoint => {
            out = out.concat(`${testPoint.outcome} ${testPoint.testNumber} ${testPoint.description}\n`);
            testPoint.diagnostics.forEach(s => {
                out = out.concat(`# ${s}\n`);
            });
        });
        epilog === null || epilog === void 0 ? void 0 : epilog.forEach(c => {
            out = out.concat(`# ${c}\n`);
        });
        return out;
    }
    buildTapResults(testResult) {
        const tapResults = [];
        testResult.tests.forEach((test, index) => {
            const testNumber = index + 1;
            const outcome = test.outcome === "Pass" /* Pass */ ? 'ok' : 'not ok';
            tapResults.push({
                testNumber,
                description: test.fullName,
                diagnostics: this.buildTapDiagnostics(test),
                outcome
            });
        });
        return tapResults;
    }
    buildTapDiagnostics(testResult) {
        const message = [];
        if (testResult.outcome !== 'Pass') {
            if (testResult.message) {
                const startsWithNewlineRegex = new RegExp(/^[/\r\n|\r|\n][\w]*/gim);
                if (startsWithNewlineRegex.test(testResult.message)) {
                    testResult.message.split(/\r\n|\r|\n/g).forEach(msg => {
                        if (msg && msg.length > 0) {
                            message.push(msg.trim());
                        }
                    });
                }
                else {
                    message.push(testResult.message);
                }
            }
            else {
                message.push('Unknown error');
            }
            if (testResult.stackTrace) {
                testResult.stackTrace.split('\n').forEach(line => {
                    message.push(line);
                });
            }
        }
        return message;
    }
}
exports.TapReporter = TapReporter;
