// @salesforce/plugin-templates/lib/commands/force/staticresource/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const staticResourceGenerator_1 = require("@salesforce/templates/lib/generators/staticResourceGenerator");
const utils_1 = require("../../../utils");
class StaticResource extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.runGenerator(staticResourceGenerator_1.default);
        });
    }
}
exports.default = StaticResource;
StaticResource.description = utils_1.MessageUtil.buildDescription('StaticResourceDescription', false);
StaticResource.examples = [
    '$ sfdx force:staticresource:create -n MyResource',
    '$ sfdx force:staticresource:create -n MyResource --contenttype application/json',
    '$ sfdx force:staticresource:create -n MyResource -d staticresources'
];
StaticResource.help = utils_1.MessageUtil.buildHelpText(StaticResource.examples, false);
StaticResource.longDescription = utils_1.MessageUtil.get('StaticResourceLongDescription');
StaticResource.flagsConfig = {
    resourcename: command_1.flags.string({
        char: 'n',
        description: utils_1.MessageUtil.get('StaticResourceNameFlagDescription'),
        longDescription: utils_1.MessageUtil.get('StaticResourceNameFlagLongDescription'),
        required: true
    }),
    contenttype: command_1.flags.string({
        description: utils_1.MessageUtil.get('StaticResourceContentTypeFlagDescription'),
        longDescription: utils_1.MessageUtil.get('StaticResourceContentTypeFlagLongDescription'),
        default: 'application/zip'
    }),
    outputdir: command_1.flags.string({
        char: 'd',
        description: utils_1.MessageUtil.get('OutputDirFlagDescription'),
        longDescription: utils_1.MessageUtil.get('OutputDirFlagLongDescription'),
        default: '.'
    }),
    apiversion: command_1.flags.builtin()
};
