// @salesforce/sfdx-plugin-lwc-test/lib/commands/force/lightning/lwc/test/run.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const cp = require("child_process");
const fs = require("fs");
const path = require("path");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/sfdx-plugin-lwc-test', 'run');
class Run extends command_1.SfdxCommand {
    // eslint-disable-next-line @typescript-eslint/require-await
    async run() {
        const args = [];
        if (this.flags.debug) {
            args.push('--debug');
        }
        else if (this.flags.watch) {
            args.push('--watch');
        }
        if (this.args.passthrough) {
            args.push(this.args.passthrough);
        }
        const scriptRet = this.runJest(args);
        this.ux.log(messages.getMessage('logSuccess', [scriptRet.status.toString()]));
        return {
            message: messages.getMessage('logSuccess', [scriptRet.status.toString()]),
            jestExitCode: scriptRet.status,
        };
    }
    runJest(args) {
        // on windows we must execute with the node prefix
        const executable = process.platform === 'win32' ? `node ${this.getExecutablePath()}` : this.getExecutablePath();
        return cp.spawnSync(executable, args, {
            stdio: 'inherit',
            shell: true,
        });
    }
    getExecutablePath() {
        const projectPath = this.project.getPath();
        const nodeModulePath = process.platform === 'win32'
            ? path.join('@salesforce', 'sfdx-lwc-jest', 'bin', 'sfdx-lwc-jest')
            : path.join('.bin', 'sfdx-lwc-jest');
        const executablePath = path.join(projectPath, 'node_modules', nodeModulePath);
        if (!fs.existsSync(executablePath)) {
            throw new core_1.SfError(messages.getMessage('errorNoExecutableFound'));
        }
        return executablePath;
    }
}
exports.default = Run;
Run.description = messages.getMessage('commandDescription');
Run.longDescription = messages.getMessage('longDescription');
Run.examples = [messages.getMessage('example1'), messages.getMessage('example2')];
Run.args = [{ name: 'passthrough' }];
// Set this to true if your command requires a project workspace; 'requiresProject' is false by default
Run.requiresProject = true;
Run.flagsConfig = {
    debug: command_1.flags.boolean({
        char: 'd',
        description: messages.getMessage('debugFlagDescription'),
        longDescription: messages.getMessage('debugFlagLongDescription'),
        exclusive: ['watch'],
    }),
    watch: command_1.flags.boolean({
        description: messages.getMessage('watchFlagDescription'),
        longDescription: messages.getMessage('watchFlagLongDescription'),
        exclusive: ['debug'],
    }),
};
