// salesforce-alm/dist/lib/org/sandbox/sandboxConstants.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandboxStatus = exports.SandboxEventNames = exports.SandboxConstants = void 0;
const kit_1 = require("@salesforce/kit");
class SandboxConstants {
}
exports.SandboxConstants = SandboxConstants;
SandboxConstants.DEFAULT_MAX_RETRIES = 0;
SandboxConstants.DEFAULT_POLL_INTERVAL = kit_1.Duration.seconds(30);
SandboxConstants.SANDBOX_INCOMPLETE_EXCEPTION_MESSAGE = 'INVALID_STATUS';
class SandboxEventNames {
}
exports.SandboxEventNames = SandboxEventNames;
SandboxEventNames.EVENT_STATUS = 'status';
SandboxEventNames.EVENT_ASYNCRESULT = 'asyncResult';
SandboxEventNames.EVENT_RESULT = 'result';
SandboxEventNames.EVENT_CLIENTID_NOTSUPPORT = 'clientIdNotSupported';
var SandboxStatus;
(function (SandboxStatus) {
    SandboxStatus["Pending"] = "Pending";
    SandboxStatus["PendingRemote"] = "Pending Remote Creation";
    SandboxStatus["Processing"] = "Processing";
    SandboxStatus["Completed"] = "Completed";
    SandboxStatus["Activating"] = "Activating";
})(SandboxStatus = exports.SandboxStatus || (exports.SandboxStatus = {}));
