// jsforce/lib/browser/jsonp.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

require("core-js/modules/es.promise");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _stringify = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/json/stringify"));

var _setTimeout2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/set-timeout"));

var _promise = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/promise"));

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _stream = require("stream");

/**
 *
 */
let _index = 0;

async function processJsonpRequest(params, jsonpParam, timeout) {
  if (params.method.toUpperCase() !== 'GET') {
    throw new Error('JSONP only supports GET request.');
  }

  _index += 1;
  const cbFuncName = `_jsforce_jsonpCallback_${_index}`;
  const callbacks = window;
  let url = params.url;
  url += (0, _indexOf.default)(url).call(url, '?') > 0 ? '&' : '?';
  url += `${jsonpParam}=${cbFuncName}`;
  const script = document.createElement('script');
  script.type = 'text/javascript';
  script.src = url;

  if (document.documentElement) {
    document.documentElement.appendChild(script);
  }

  let pid;

  try {
    const res = await new _promise.default((resolve, reject) => {
      pid = (0, _setTimeout2.default)(() => {
        reject(new Error('JSONP call time out.'));
      }, timeout);
      callbacks[cbFuncName] = resolve;
    });
    return {
      statusCode: 200,
      headers: {
        'content-type': 'application/json'
      },
      body: (0, _stringify.default)(res)
    };
  } finally {
    clearTimeout(pid);

    if (document.documentElement) {
      document.documentElement.removeChild(script);
    }

    delete callbacks[cbFuncName];
  }
}

function createRequest(jsonpParam = 'callback', timeout = 10000) {
  return params => {
    const stream = new _stream.Transform({
      transform(chunk, encoding, callback) {
        callback();
      },

      flush() {
        (async () => {
          const response = await processJsonpRequest(params, jsonpParam, timeout);
          stream.emit('response', response);
          stream.emit('complete', response);
          stream.push(response.body);
          stream.push(null);
        })();
      }

    });
    stream.end();
    return stream;
  };
}

var _default = {
  supported: typeof window !== 'undefined' && typeof document !== 'undefined',
  createRequest
};
exports.default = _default;
