// jsforce/module/browser/canvas.js
import "core-js/modules/es.array.iterator";
import "core-js/modules/es.promise";
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";

/**
 *
 */
import { Transform } from 'stream';

function parseHeaders(hs) {
  const headers = {};

  for (const line of hs.split(/\n/)) {
    const [name, value] = line.split(/\s*:\s*/);
    headers[name.toLowerCase()] = value;
  }

  return headers;
}

async function processCanvasRequest(params, signedRequest, requestBody) {
  const settings = {
    client: signedRequest.client,
    method: params.method,
    data: requestBody
  };
  const paramHeaders = params.headers;

  if (paramHeaders) {
    settings.headers = {};

    for (const name of _Object$keys(paramHeaders)) {
      if (name.toLowerCase() === 'content-type') {
        settings.contentType = paramHeaders[name];
      } else {
        settings.headers[name] = paramHeaders[name];
      }
    }
  }

  const data = await new _Promise((resolve, reject) => {
    settings.success = resolve;
    settings.failure = reject;
    Sfdc.canvas.client.ajax(params.url, settings);
  });
  const headers = parseHeaders(data.responseHeaders);
  let responseBody = data.payload;

  if (typeof responseBody !== 'string') {
    responseBody = _JSON$stringify(responseBody);
  }

  return {
    statusCode: data.status,
    headers,
    body: responseBody
  };
}

function createRequest(signedRequest) {
  return params => {
    const buf = [];
    const stream = new Transform({
      transform(chunk, encoding, callback) {
        buf.push(typeof chunk === 'string' ? chunk : chunk.toString('utf8'));
        callback();
      },

      flush() {
        (async () => {
          const body = buf.join('');
          const response = await processCanvasRequest(params, signedRequest, body);
          stream.emit('response', response);
          stream.emit('complete', response);
          stream.push(response.body);
          stream.push(null);
        })();
      }

    });

    if (params.body) {
      stream.end(params.body);
    }

    return stream;
  };
}

export default {
  supported: typeof Sfdc === 'object' && typeof Sfdc.canvas !== 'undefined',
  createRequest
};
