// salesforce-alm/dist/lib/source/workspaceFileState.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.toReadableState = exports.WorkspaceFileState = void 0;
var WorkspaceFileState;
(function (WorkspaceFileState) {
    WorkspaceFileState["UNCHANGED"] = "u";
    WorkspaceFileState["CHANGED"] = "c";
    WorkspaceFileState["DELETED"] = "d";
    WorkspaceFileState["NEW"] = "n";
    WorkspaceFileState["DUP"] = "p";
})(WorkspaceFileState = exports.WorkspaceFileState || (exports.WorkspaceFileState = {}));
function toReadableState(state) {
    switch (state) {
        case WorkspaceFileState.UNCHANGED:
            return 'Unchanged';
        case WorkspaceFileState.CHANGED:
            return 'Changed';
        case WorkspaceFileState.DELETED:
            return 'Deleted';
        case WorkspaceFileState.NEW:
            return 'Add';
        case WorkspaceFileState.DUP:
            return 'Duplicate';
        default:
            return 'Unknown';
    }
}
exports.toReadableState = toReadableState;
