// salesforce-alm/messages/mdapi_retrieve.js
module.exports = {
  retrieveCommandDescription: 'retrieve metadata from an org using Metadata API',
  retrieveCommandLongDescription:
    'Uses Metadata API to retrieve a .zip of XML files that represent metadata from ' +
    'the targeted org. The default target username is the admin user for the default scratch org. You can ' +
    'retrieve and deploy up to 10,000 files or 400 MB (39 MB compressed) at one time.',
  retrieveCommandHelp:
    'The default target username is the admin user for the default scratch org. You can ' +
    'retrieve and deploy up to 10,000 files or 400 MB (39 MB compressed) at one time.',
  reportCommandDescription: 'check the status of a metadata retrieval',
  reportCommandLongDescription: 'Check the status of an asynchronous metadata retrieval.',
  reportCommandHelp:
    'Specify the job ID and a target directory for the retrieve you want to check. You can also ' +
    'specify a wait time (minutes) to check for updates to the deploy status. If the retrieve was successful, ' +
    'the resulting zip file will be saved to the location passed in with the retrieve target parameter.',
  retrievetargetdirFlagDescription: 'directory root for the retrieved files',
  retrievetargetdirFlagLongDescription:
    'The root of the directory structure where the retrieved .zip or metadata files are put.',
  unpackagedFlagDescription: 'file path of manifest of components to retrieve',
  unpackagedFlagLongDescription: 'The complete path for the manifest file that specifies the components to retrieve.',
  verboseFlagDescription: 'verbose output of retrieve result',
  verboseFlagLongDescription: 'Indicates that you want verbose output from the retrieve operation.',
  sourcedirFlagDescription: 'source dir to use instead of the default package dir in %s',
  sourcedirFlagLongDescription: 'The source directory to use instead of the default package directory specified in %s',
  singlepackageFlagDescription: 'Indicates that the zip file points to a directory structure for a single package',
  singlepackageFlagLongDescription:
    'Indicates that the specified .zip file points to a directory structure for a single package. By default, the CLI assumes the directory is structured for a set of packages.',
  packagenamesFlagDescription: 'a comma-separated list of packages to retrieve',
  packagenamesFlagLongDescription: 'A comma-separated list of package names to retrieve.',
  apiversionFlagDescription: 'target API version for the retrieve',
  apiversionFlagLongDescription:
    'Use to override the default, which is the latest version supported by ' +
    'your CLI plug-in, with the version in your package.xml file.',
  waitFlagDescription: 'wait time for command to finish in minutes (default: %s (no limit))',
  waitFlagLongDescription: 'The number of minutes to wait for the command to complete.',
  jobidDescription:
    'job ID of the retrieve you want to check; defaults to your most recent CLI retrieval if not specified',
  jobidLongDescription:
    'The job ID (asyncId) of the retrieve you want to check. If not specified, the default value is the ID' +
    ' of the most recent metadata retrieval you ran using Salesforce CLI. You must specify a ' +
    '--retrievetargetdir. Use with --wait to resume waiting.',
};
