// salesforce-alm/messages/source_deploy_cancel.js
module.exports = {
  description: 'cancel a source deployment',
  longDescription: 'Cancels an asynchronous source deployment.',
  help: `Use this command to cancel a specified asynchronous source deployment. You can also specify a wait time (in minutes) to check for updates to the canceled deploy status.

To run the command asynchronously, set --wait to 0, which immediately returns the job ID. This way, you can continue to use the CLI.
To check the status of the job, use force:source:deploy:report.

Examples:

Deploy a directory of files to the org
  $ sfdx force:source:deploy -d <directory>
Now cancel this deployment and wait two minutes
  $ sfdx force:source:deploy:cancel -w 2

If you have multiple deployments in progress and want to cancel a specific one, specify the job ID
  $ sfdx force:source:deploy:cancel -i <jobid>
Check the status of the cancel job
  $ sfdx force:source:deploy:report`,

  jobId: 'job ID of the deployment you want to cancel; defaults to your most recent CLI deployment if not specified',
  jobIdLong: 'The job ID (requestId) of the deployment you want to cancel. If not specified, the default value is the ID of the most recent source deployment you ran using Salesforce CLI.',

  MissingRequiredParameter: 'A required parameter [jobid] is missing.',
  CancelFailed: 'The cancel command failed due to: %s'
};
