// salesforce-alm/messages/source_push.js
module.exports = {
  pushCommandConflictMsg:
    "We couldn't complete the push operation due to conflicts. " +
    'Verify that you want to keep the local versions, then run "sfdx force:source:legacy:push -f" with the --forceoverwrite (-f) option.',
  commandDescription: 'push source to a scratch org from the project',
  commandDescriptionLong: 'Pushes changed source from your project to a scratch org to keep them in sync.',
  commandHelp:
    'If the command detects a conflict, it displays the conflicts but does not complete the process. ' +
    'After reviewing the conflict, rerun the command with the --forceoverwrite parameter.',
  waitFlagDescriptionLong:
    'Number of minutes to wait for the command to complete and display ' +
    'results to the terminal window. If the command continues to run after the wait period, the CLI ' +
    'returns control of the terminal window to you. The default is 33 minutes.',
  forceoverwriteFlagDescription: 'ignore conflict warnings and overwrite changes to scratch org',
  forceoverwriteFlagDescriptionLong:
    'Runs the push command even if conflicts exist. Changes in the project overwrite changes in the scratch org.',
  replacetokensFlagDescription: 'replace tokens in source files prior to deployment',
  replacetokensFlagDescriptionLong: 'Replaces tokens in source files prior to deployment.',
  ignorewarningsFlagDescription: 'deploy changes even if warnings are generated',
  ignorewarningsFlagDescriptionLong: 'Completes the deployment even if warnings are generated.',
  sourcepushFailed: 'Push failed.',
};
