// @salesforce/source-deploy-retrieve/lib/src/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deepFreeze = exports.normalizeToArray = exports.parseNestedFullName = exports.trimUntil = exports.parentName = exports.parseMetadataXml = exports.baseName = exports.extName = exports.trimMetaXmlSuffix = exports.generateMetaXMLPath = exports.generateMetaXML = exports.createFiles = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var fileSystemHandler_1 = require("./fileSystemHandler");
Object.defineProperty(exports, "createFiles", { enumerable: true, get: function () { return fileSystemHandler_1.createFiles; } });
var metadata_1 = require("./metadata");
Object.defineProperty(exports, "generateMetaXML", { enumerable: true, get: function () { return metadata_1.generateMetaXML; } });
Object.defineProperty(exports, "generateMetaXMLPath", { enumerable: true, get: function () { return metadata_1.generateMetaXMLPath; } });
Object.defineProperty(exports, "trimMetaXmlSuffix", { enumerable: true, get: function () { return metadata_1.trimMetaXmlSuffix; } });
var path_1 = require("./path");
Object.defineProperty(exports, "extName", { enumerable: true, get: function () { return path_1.extName; } });
Object.defineProperty(exports, "baseName", { enumerable: true, get: function () { return path_1.baseName; } });
Object.defineProperty(exports, "parseMetadataXml", { enumerable: true, get: function () { return path_1.parseMetadataXml; } });
Object.defineProperty(exports, "parentName", { enumerable: true, get: function () { return path_1.parentName; } });
Object.defineProperty(exports, "trimUntil", { enumerable: true, get: function () { return path_1.trimUntil; } });
Object.defineProperty(exports, "parseNestedFullName", { enumerable: true, get: function () { return path_1.parseNestedFullName; } });
var collections_1 = require("./collections");
Object.defineProperty(exports, "normalizeToArray", { enumerable: true, get: function () { return collections_1.normalizeToArray; } });
Object.defineProperty(exports, "deepFreeze", { enumerable: true, get: function () { return collections_1.deepFreeze; } });
