// jsforce/lib/api/apex.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = exports.Apex = void 0;

var _stringify = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/json/stringify"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/defineProperty"));

var _jsforce = require("../jsforce");

/**
 * @file Manages Salesforce Apex REST endpoint calls
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */

/**
 * API class for Apex REST endpoint call
 */
class Apex {
  /**
   *
   */
  constructor(conn) {
    (0, _defineProperty2.default)(this, "_conn", void 0);
    (0, _defineProperty2.default)(this, "del", this.delete);
    this._conn = conn;
  }
  /* @private */


  _baseUrl() {
    return `${this._conn.instanceUrl}/services/apexrest`;
  }
  /**
   * @private
   */


  _createRequestParams(method, path, body, options = {}) {
    const headers = typeof options.headers === 'object' ? options.headers : {};

    if (!/^(GET|DELETE)$/i.test(method)) {
      headers['content-type'] = 'application/json';
    }

    const params = {
      method,
      url: this._baseUrl() + path,
      headers
    };

    if (body) {
      params.body = (0, _stringify.default)(body);
    }

    return params;
  }
  /**
   * Call Apex REST service in GET request
   */


  get(path, options) {
    return this._conn.request(this._createRequestParams('GET', path, undefined, options));
  }
  /**
   * Call Apex REST service in POST request
   */


  post(path, body, options) {
    const params = this._createRequestParams('POST', path, body, options);

    return this._conn.request(params);
  }
  /**
   * Call Apex REST service in PUT request
   */


  put(path, body, options) {
    const params = this._createRequestParams('PUT', path, body, options);

    return this._conn.request(params);
  }
  /**
   * Call Apex REST service in PATCH request
   */


  patch(path, body, options) {
    const params = this._createRequestParams('PATCH', path, body, options);

    return this._conn.request(params);
  }
  /**
   * Call Apex REST service in DELETE request
   */


  delete(path, options) {
    return this._conn.request(this._createRequestParams('DELETE', path, undefined, options));
  }
  /**
   * Synonym of Apex#delete()
   */


}
/*--------------------------------------------*/

/*
 * Register hook in connection instantiation for dynamically adding this API module features
 */


exports.Apex = Apex;
(0, _jsforce.registerModule)('apex', conn => new Apex(conn));
var _default = Apex;
exports.default = _default;
