// @salesforce/plugin-source/lib/formatters/mdapi/mdDeployResultFormatter.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MdDeployResultFormatter = void 0;
const chalk = require("chalk");
const ts_types_1 = require("@salesforce/ts-types");
const core_1 = require("@salesforce/core");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
const resultFormatter_1 = require("../resultFormatter");
const coverageUtils_1 = require("../../coverageUtils");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-source', 'md.deploy');
class MdDeployResultFormatter extends resultFormatter_1.ResultFormatter {
    constructor(logger, ux, options, result) {
        super(logger, ux, options);
        this.result = result;
    }
    /**
     * Get the JSON output from the DeployResult.
     *
     * @returns a JSON formatted result matching the provided type.
     */
    getJson() {
        // concise omits success messages
        // spread properties prevents modification of the object from impacting tests
        const response = this.getResponse();
        if (this.options.testsRan) {
            response.coverage = this.getCoverageFileInfo();
            response.junit = this.getJunitFileInfo();
        }
        if (this.isConcise()) {
            return {
                ...response,
                details: {
                    componentFailures: response.details.componentFailures,
                    runTestResult: response.details.runTestResult,
                },
            };
        }
        return response;
    }
    /**
     * @param isReportCommand set to true if the command is deploy:report.  The output is slightly different
     * Displays deploy results in human readable format.  Output can vary based on:
     *
     * 1. Verbose option
     * 3. Checkonly deploy (checkonly=true)
     * 4. Deploy with test results
     * 5. Canceled status
     */
    display(isReportCommand = false) {
        var _a, _b;
        if (this.hasStatus(source_deploy_retrieve_1.RequestStatus.Canceled)) {
            const canceledByName = (_a = this.result.response.canceledByName) !== null && _a !== void 0 ? _a : 'unknown';
            throw new core_1.SfError(messages.getMessage('deployCanceled', [canceledByName]), 'DeployFailed');
        }
        if (this.isVerbose()) {
            this.displaySuccesses();
            this.displayFailures();
            this.displayTestResults(isReportCommand);
            this.displayOutputFileLocations();
        }
        else if (isReportCommand) {
            this.ux.log(`Status: ${(_b = this.result.response.status) !== null && _b !== void 0 ? _b : 'unknown'}`);
            const deploys = `Deployed: ${this.getNumResult('numberComponentsDeployed')}/${this.getNumResult('numberComponentsTotal')}`;
            const deployErrors = `Errors: ${this.getNumResult('numberComponentErrors')}`;
            const tests = `Tests Complete: ${this.getNumResult('numberTestsCompleted')}/${this.getNumResult('numberTestsTotal')}`;
            const testErrs = `Errors: ${this.getNumResult('numberTestErrors')}`;
            this.ux.log(`${deploys} ${deployErrors}`);
            this.ux.log(`${tests} ${testErrs}`);
            this.displayOutputFileLocations();
        }
        else {
            // always show failures
            this.displayFailures();
            this.displayTestResults(isReportCommand);
            this.displayOutputFileLocations();
        }
        // TODO: the toolbelt version of this is returning an SfError shape.  This returns a status=1 and the result (mdapi response) but not the error name, etc
        if (!this.isSuccess()) {
            const error = new core_1.SfError(messages.getMessage('deployFailed'), 'mdapiDeployFailed');
            error.setData(this.result);
            throw error;
        }
    }
    hasStatus(status) {
        return this.result.response.status === status;
    }
    getResponse() {
        var _a;
        return (_a = this.result.response) !== null && _a !== void 0 ? _a : {};
    }
    displaySuccesses() {
        if (this.isSuccess() && this.isVerbose()) {
            const successes = (0, resultFormatter_1.toArray)(this.getResponse().details.componentSuccesses).sort(mdResponseSorter);
            this.ux.log('');
            this.ux.styledHeader(chalk.blue('Deployed Source'));
            this.ux.table(successes, {
                componentType: { header: 'Type' },
                fileName: { header: 'File' },
                fullName: { header: 'Name' },
                id: { header: 'Id' },
            });
        }
    }
    displayFailures() {
        if (this.hasStatus(source_deploy_retrieve_1.RequestStatus.Failed) || this.hasStatus(source_deploy_retrieve_1.RequestStatus.SucceededPartial)) {
            const failures = (0, resultFormatter_1.toArray)(this.getResponse().details.componentFailures).sort(mdResponseSorter);
            if (failures.length === 0) {
                return;
            }
            this.ux.log('');
            this.ux.styledHeader(chalk.red(`Component Failures [${failures.length}]`));
            this.ux.table(failures, {
                problemType: { header: 'Type' },
                fileName: { header: 'File' },
                fullName: { header: 'Name' },
                problem: { header: 'Problem' },
            });
            this.ux.log('');
        }
    }
    getNumResult(field) {
        return (0, ts_types_1.getNumber)(this.result, `response.${field}`, 0);
    }
    displayTestResults(isReportCommand = false) {
        if (this.result.response.runTestsEnabled) {
            this.ux.log('');
            if (this.isVerbose() || !isReportCommand) {
                this.verboseTestFailures();
                this.verboseTestSuccess();
                this.verboseTestTime();
            }
            else {
                this.ux.styledHeader(chalk.blue('Test Results Summary'));
                this.ux.log(`Passing: ${this.getNumResult('numberTestsCompleted')}`);
                this.ux.log(`Failing: ${this.getNumResult('numberTestErrors')}`);
                this.ux.log(`Total: ${this.getNumResult('numberTestsTotal')}`);
                this.ux.log(`Time: ${this.getNumResult('details.runTestResult.totalTime')}`);
            }
        }
    }
    verboseTestFailures() {
        var _a, _b, _c, _d, _e;
        if ((_b = (_a = this.result) === null || _a === void 0 ? void 0 : _a.response) === null || _b === void 0 ? void 0 : _b.numberTestErrors) {
            const failures = (0, resultFormatter_1.toArray)((_d = (_c = this.result.response.details) === null || _c === void 0 ? void 0 : _c.runTestResult) === null || _d === void 0 ? void 0 : _d.failures);
            const tests = this.sortTestResults(failures);
            this.ux.log('');
            this.ux.styledHeader(chalk.red(`Test Failures [${(_e = this.result.response.details.runTestResult) === null || _e === void 0 ? void 0 : _e.numFailures}]`));
            this.ux.table(tests, {
                name: { header: 'Name' },
                methodName: { header: 'Method' },
                message: { header: 'Message' },
                stackTrace: { header: 'Stacktrace' },
            });
        }
    }
    verboseTestSuccess() {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const success = (0, resultFormatter_1.toArray)((_d = (_c = (_b = (_a = this.result) === null || _a === void 0 ? void 0 : _a.response) === null || _b === void 0 ? void 0 : _b.details) === null || _c === void 0 ? void 0 : _c.runTestResult) === null || _d === void 0 ? void 0 : _d.successes);
        if (success.length) {
            const tests = this.sortTestResults(success);
            this.ux.log('');
            this.ux.styledHeader(chalk.green(`Test Success [${success.length}]`));
            this.ux.table(tests, {
                name: { header: 'Name' },
                methodName: { header: 'Method' },
            });
        }
        const codeCoverage = (0, resultFormatter_1.toArray)((_h = (_g = (_f = (_e = this.result) === null || _e === void 0 ? void 0 : _e.response) === null || _f === void 0 ? void 0 : _f.details) === null || _g === void 0 ? void 0 : _g.runTestResult) === null || _h === void 0 ? void 0 : _h.codeCoverage);
        if (codeCoverage.length) {
            const coverage = (0, coverageUtils_1.prepCoverageForDisplay)(codeCoverage);
            this.ux.log('');
            this.ux.styledHeader(chalk.blue('Apex Code Coverage'));
            this.ux.table(coverage, {
                name: { header: 'Name' },
                numLocations: { header: '% Covered' },
                lineNotCovered: { header: 'Uncovered Lines' },
            });
        }
    }
    verboseTestTime() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        if (((_c = (_b = (_a = this.result.response) === null || _a === void 0 ? void 0 : _a.details) === null || _b === void 0 ? void 0 : _b.runTestResult) === null || _c === void 0 ? void 0 : _c.successes) ||
            ((_g = (_f = (_e = (_d = this.result) === null || _d === void 0 ? void 0 : _d.response) === null || _e === void 0 ? void 0 : _e.details) === null || _f === void 0 ? void 0 : _f.runTestResult) === null || _g === void 0 ? void 0 : _g.failures)) {
            this.ux.log('');
            this.ux.log(`Total Test Time:  ${(_l = (_k = (_j = (_h = this.result) === null || _h === void 0 ? void 0 : _h.response) === null || _j === void 0 ? void 0 : _j.details) === null || _k === void 0 ? void 0 : _k.runTestResult) === null || _l === void 0 ? void 0 : _l.totalTime}`);
        }
    }
}
exports.MdDeployResultFormatter = MdDeployResultFormatter;
const mdResponseSorter = (i, j) => {
    if (i.componentType === j.componentType) {
        if (i.fileName === j.fileName) {
            return i.fullName > j.fullName ? 1 : -1;
        }
        return i.fileName > j.fileName ? 1 : -1;
    }
    return i.componentType > j.componentType ? 1 : -1;
};
