// @salesforce/core/lib/config/sandboxOrgConfig.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandboxOrgConfig = void 0;
const configFile_1 = require("./configFile");
/**
 * A config file that stores usernames for an org.
 */
class SandboxOrgConfig extends configFile_1.ConfigFile {
    /**
     * Constructor.
     * **Do not directly construct instances of this class -- use {@link SandboxConfig.create} instead.**
     *
     * @param options The options for the class instance
     * @ignore
     */
    constructor(options) {
        super(options);
    }
    /**
     * Gets the config options for a given org ID.
     *
     * @param orgId The orgId. Generally this org would have multiple users configured.
     */
    static getOptions(orgId) {
        return {
            isGlobal: true,
            isState: true,
            filename: `${orgId}.sandbox.json`,
            orgId,
        };
    }
}
exports.SandboxOrgConfig = SandboxOrgConfig;
(function (SandboxOrgConfig) {
    let Fields;
    (function (Fields) {
        /**
         * The username of the user who created the sandbox.
         */
        Fields["PROD_ORG_USERNAME"] = "prodOrgUsername";
    })(Fields = SandboxOrgConfig.Fields || (SandboxOrgConfig.Fields = {}));
})(SandboxOrgConfig = exports.SandboxOrgConfig || (exports.SandboxOrgConfig = {}));
