// jsforce/browser/browser/canvas.js
import _getIterator from "@babel/runtime-corejs3/core-js/get-iterator";
import _Array$isArray from "@babel/runtime-corejs3/core-js-stable/array/is-array";
import _getIteratorMethod from "@babel/runtime-corejs3/core-js/get-iterator-method";
import _Symbol from "@babel/runtime-corejs3/core-js-stable/symbol";
import "core-js/modules/es.array.join";
import "core-js/modules/es.function.name";
import "core-js/modules/es.object.to-string";
import "core-js/modules/es.regexp.exec";
import "core-js/modules/es.regexp.to-string";
import "core-js/modules/es.string.split";
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import _Object$keys2 from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _Array$from from "@babel/runtime-corejs3/core-js-stable/array/from";
import _sliceInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/slice";
import _typeof from "@babel/runtime-corejs3/helpers/typeof";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
import "regenerator-runtime/runtime";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime-corejs3/helpers/slicedToArray";

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof _Symbol === "undefined" || _getIteratorMethod(o) == null) { if (_Array$isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = _getIterator(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { var _context2; if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = _sliceInstanceProperty(_context2 = Object.prototype.toString.call(o)).call(_context2, 8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return _Array$from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

/**
 *
 */
import { Transform } from 'stream';

function parseHeaders(hs) {
  var headers = {};

  var _iterator = _createForOfIteratorHelper(hs.split(/\n/)),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var line = _step.value;

      var _line$split = line.split(/\s*:\s*/),
          _line$split2 = _slicedToArray(_line$split, 2),
          name = _line$split2[0],
          value = _line$split2[1];

      headers[name.toLowerCase()] = value;
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return headers;
}

function processCanvasRequest(_x, _x2, _x3) {
  return _processCanvasRequest.apply(this, arguments);
}

function _processCanvasRequest() {
  _processCanvasRequest = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(params, signedRequest, requestBody) {
    var settings, paramHeaders, _i, _Object$keys, name, data, headers, responseBody;

    return _regeneratorRuntime.wrap(function _callee2$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            settings = {
              client: signedRequest.client,
              method: params.method,
              data: requestBody
            };
            paramHeaders = params.headers;

            if (paramHeaders) {
              settings.headers = {};

              for (_i = 0, _Object$keys = _Object$keys2(paramHeaders); _i < _Object$keys.length; _i++) {
                name = _Object$keys[_i];

                if (name.toLowerCase() === 'content-type') {
                  settings.contentType = paramHeaders[name];
                } else {
                  settings.headers[name] = paramHeaders[name];
                }
              }
            }

            _context3.next = 5;
            return new _Promise(function (resolve, reject) {
              settings.success = resolve;
              settings.failure = reject;
              Sfdc.canvas.client.ajax(params.url, settings);
            });

          case 5:
            data = _context3.sent;
            headers = parseHeaders(data.responseHeaders);
            responseBody = data.payload;

            if (typeof responseBody !== 'string') {
              responseBody = _JSON$stringify(responseBody);
            }

            return _context3.abrupt("return", {
              statusCode: data.status,
              headers: headers,
              body: responseBody
            });

          case 10:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee2);
  }));
  return _processCanvasRequest.apply(this, arguments);
}

function createRequest(signedRequest) {
  return function (params) {
    var buf = [];
    var stream = new Transform({
      transform: function transform(chunk, encoding, callback) {
        buf.push(typeof chunk === 'string' ? chunk : chunk.toString('utf8'));
        callback();
      },
      flush: function flush() {
        _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
          var body, response;
          return _regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  body = buf.join('');
                  _context.next = 3;
                  return processCanvasRequest(params, signedRequest, body);

                case 3:
                  response = _context.sent;
                  stream.emit('response', response);
                  stream.emit('complete', response);
                  stream.push(response.body);
                  stream.push(null);

                case 8:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee);
        }))();
      }
    });

    if (params.body) {
      stream.end(params.body);
    }

    return stream;
  };
}

export default {
  supported: (typeof Sfdc === "undefined" ? "undefined" : _typeof(Sfdc)) === 'object' && typeof Sfdc.canvas !== 'undefined',
  createRequest: createRequest
};
