// @salesforce/plugin-source/lib/commands/force/mdapi/deploy/cancel.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cancel = void 0;
const os = require("os");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
const deployCommand_1 = require("../../../../deployCommand");
const deployCancelResultFormatter_1 = require("../../../../formatters/deployCancelResultFormatter");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.load('@salesforce/plugin-source', 'md.cancel', [
    'CancelFailed',
    'description',
    'examples',
    'flags.wait',
    'flags.waitLong',
    'flags.jobid',
]);
class Cancel extends deployCommand_1.DeployCommand {
    async run() {
        await this.cancel();
        this.resolveSuccess();
        return this.formatResult();
    }
    async cancel() {
        const deployId = this.resolveDeployId(this.getFlag('jobid'));
        try {
            const deploy = this.createDeploy(deployId);
            await deploy.cancel();
            this.deployResult = await this.poll(deployId);
        }
        catch (e) {
            const err = e;
            throw new core_1.SfError(messages.getMessage('CancelFailed', [err.message]), 'CancelFailed');
        }
    }
    resolveSuccess() {
        const status = this.deployResult.response.status;
        if (status !== source_deploy_retrieve_1.RequestStatus.Canceled) {
            this.setExitCode(1);
        }
    }
    formatResult() {
        const formatter = new deployCancelResultFormatter_1.DeployCancelResultFormatter(this.logger, this.ux, this.deployResult);
        if (!this.isJsonOutput()) {
            formatter.display();
        }
        return formatter.getJson();
    }
}
exports.Cancel = Cancel;
Cancel.description = messages.getMessage('description');
Cancel.examples = messages.getMessage('examples').split(os.EOL);
Cancel.requiresUsername = true;
Cancel.flagsConfig = {
    wait: command_1.flags.minutes({
        char: 'w',
        default: kit_1.Duration.minutes(deployCommand_1.DeployCommand.DEFAULT_WAIT_MINUTES),
        min: kit_1.Duration.minutes(1),
        description: messages.getMessage('flags.wait'),
        longDescription: messages.getMessage('flags.waitLong'),
    }),
    jobid: command_1.flags.id({
        char: 'i',
        description: messages.getMessage('flags.jobid'),
        validate: deployCommand_1.DeployCommand.isValidDeployId,
    }),
};
