// @salesforce/core/lib/config/sandboxProcessCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandboxRequestCache = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const kit_1 = require("@salesforce/kit");
const global_1 = require("../global");
const ttlConfig_1 = require("./ttlConfig");
class SandboxRequestCache extends ttlConfig_1.TTLConfig {
    static getDefaultOptions() {
        return {
            isGlobal: true,
            isState: true,
            filename: SandboxRequestCache.getFileName(),
            stateFolder: global_1.Global.SF_STATE_FOLDER,
            ttl: kit_1.Duration.days(14),
        };
    }
    static async unset(key) {
        const cache = await SandboxRequestCache.create();
        cache.unset(key);
        await cache.write();
    }
    static async set(key, sandboxProcessObject) {
        const cache = await SandboxRequestCache.create();
        cache.set(key, sandboxProcessObject);
        await cache.write();
    }
    static getFileName() {
        return 'sandbox-create-cache.json';
    }
}
exports.SandboxRequestCache = SandboxRequestCache;
