// @salesforce/plugin-data/lib/dataSoqlQueryTypes.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAttributesElement = exports.hasNestedRecords = exports.FieldType = void 0;
var FieldType;
(function (FieldType) {
    FieldType[FieldType["field"] = 0] = "field";
    FieldType[FieldType["subqueryField"] = 1] = "subqueryField";
    FieldType[FieldType["functionField"] = 2] = "functionField";
})(FieldType = exports.FieldType || (exports.FieldType = {}));
const hasNestedRecords = (element) => {
    return Array.isArray(element === null || element === void 0 ? void 0 : element.records);
};
exports.hasNestedRecords = hasNestedRecords;
const isAttributesElement = (element) => {
    return (!!element.referenceId && !!element.type);
};
exports.isAttributesElement = isAttributesElement;
