// @salesforce/plugin-data/lib/dataCommand.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataCommand = void 0;
const fse = require("fs-extra");
const command_1 = require("@salesforce/command");
const ts_types_1 = require("@salesforce/ts-types");
const core_1 = require("@salesforce/core");
const http_api_1 = require("jsforce/lib/http-api");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-data', 'messages');
// eslint-disable-next-line @typescript-eslint/unbound-method
const originalRequestMethod = http_api_1.HttpApi.prototype.request;
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
http_api_1.HttpApi.prototype.request = function (req, ...args) {
    this.once('response', (response) => {
        var _a;
        const metrics = response.headers.perfmetrics;
        if (metrics) {
            DataCommand.addMetric({
                requestPath: `/services${(_a = req === null || req === void 0 ? void 0 : req.url) === null || _a === void 0 ? void 0 : _a.split('/services')[1]}`,
                perfMetrics: JSON.parse(metrics),
            });
        }
    });
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return originalRequestMethod.call(this, req, ...args);
};
/* eslint-enable @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-call*/
class DataCommand extends command_1.SfdxCommand {
    static addMetric(metric) {
        DataCommand.metrics.push(metric);
    }
    static getMetrics() {
        return DataCommand.metrics;
    }
    validateIdXorWhereFlags() {
        if (!this.flags.where && !this.flags.sobjectid) {
            throw new core_1.SfError(messages.getMessage('NeitherSobjectidNorWhereError'), 'NeitherSobjectidNorWhereError', [
                messages.getMessage('NeitherSobjectidNorWhereErrorActions'),
            ]);
        }
    }
    collectErrorMessages(result) {
        let errors = '';
        if (result.errors) {
            errors = '\nErrors:\n';
            result.errors.map((err) => {
                errors += '  ' + (err === null || err === void 0 ? void 0 : err.message) + '\n';
            });
        }
        return errors;
    }
    async throwIfPathDoesntExist(path) {
        if (!(await fse.pathExists(path))) {
            throw new core_1.SfError(messages.getMessage('PathDoesNotExist', [path]), 'PathDoesNotExist');
        }
    }
    getJsonResultObject(result = this.result.data, status = process.exitCode || 0) {
        const final = { status, result };
        const perfMetrics = DataCommand.getMetrics();
        if (perfMetrics.length)
            final.perfMetrics = perfMetrics;
        return final;
    }
    /**
     * Necessary where plugin commands are extending a base class that extends SfdxCommand
     *
     * @returns Org
     */
    ensureOrg() {
        if (!this.org) {
            throw new core_1.SfError('This command requires a username. Specify it with the -u parameter or with the "sfdx config:set defaultusername=<username>" command.');
        }
        return this.org;
    }
    getConnection() {
        const safeOrg = this.ensureOrg();
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const connection = this.flags.usetoolingapi
            ? safeOrg.getConnection().tooling
            : safeOrg.getConnection();
        if (this.flags.perflog) {
            // eslint-disable-next-line no-underscore-dangle
            if (!connection._callOptions) {
                // eslint-disable-next-line no-underscore-dangle
                connection._callOptions = {};
            }
            // eslint-disable-next-line no-underscore-dangle
            connection._callOptions.perfOption = 'MINIMUM';
        }
        return connection;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore enable any typing here
    async query(sobject, where) {
        const queryObject = this.stringToDictionary(where);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
        const records = await sobject.find(queryObject, 'id');
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        if (!records || records.length === 0) {
            throw new core_1.SfError('DataRecordGetNoRecord', messages.getMessage('DataRecordGetNoRecord'));
        }
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        if (records.length > 1) {
            throw new core_1.SfError('DataRecordGetMultipleRecords', 
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            messages.getMessage('DataRecordGetMultipleRecords', [where, this.flags.sobjecttype, records.length]));
        }
        return this.normalize(records);
    }
    stringToDictionary(str) {
        const keyValuePairs = this.parseKeyValueSequence(str);
        return this.transformKeyValueSequence(keyValuePairs);
    }
    normalize(results) {
        // jsforce returns SaveResult | SaveResult[]
        // but we're only ever dealing with a single sobject we are guaranteed to
        // get back a single SaveResult. Nevertheless, we ensure that it's a
        // single SaveResult to make Typescript happy
        return Array.isArray(results) ? results[0] : results;
    }
    logNestedObject(obj, indentation = 0) {
        const space = ' '.repeat(indentation);
        Object.keys(obj).forEach((key) => {
            const value = (0, ts_types_1.get)(obj, key, null);
            if (!!value && typeof value === 'object') {
                this.ux.log(`${space}${key}:`);
                this.logNestedObject(value, indentation + 2);
            }
            else {
                this.ux.log(`${space}${key}: ${value}`);
            }
        });
    }
    /**
     * Takes a sequence of key=value string pairs and produces an object out of them.
     * If you repeat the key, it replaces the value with the subsequent value.
     *
     * @param [keyValuePairs] - The list of key=value pair strings.
     */
    transformKeyValueSequence(keyValuePairs) {
        const constructedObject = {};
        keyValuePairs.forEach((pair) => {
            // Look for the *first* '=' and splits there, ignores any subsequent '=' for this pair
            const eqPosition = pair.indexOf('=');
            if (eqPosition === -1) {
                throw new Error(messages.getMessage('TextUtilMalformedKeyValuePair', [pair]));
            }
            else {
                const key = pair.substr(0, eqPosition);
                if (pair.includes('{') && pair.includes('}')) {
                    try {
                        constructedObject[key] = JSON.parse(pair.substr(eqPosition + 1));
                    }
                    catch {
                        // the data contained { and }, but wasn't valid JSON, default to parsing as-is
                        constructedObject[key] = this.convertToBooleanIfApplicable(pair.substr(eqPosition + 1));
                    }
                }
                else {
                    constructedObject[key] = this.convertToBooleanIfApplicable(pair.substr(eqPosition + 1));
                }
            }
        });
        return constructedObject;
    }
    convertToBooleanIfApplicable(input) {
        if (input.trim().toLowerCase() === 'false')
            return false;
        if (input.trim().toLowerCase() === 'true')
            return true;
        return input;
    }
    /**
     * Splits a sequence of 'key=value key="leftValue rightValue"   key=value'
     * into a list of key=value pairs, paying attention to quoted whitespace.
     *
     * This is NOT a full push down-automaton so do NOT expect full error handling/recovery.
     *
     * @param {string} text - The sequence to split
     */
    parseKeyValueSequence(text) {
        const separator = /\s/;
        let inSingleQuote = false;
        let inDoubleQuote = false;
        let currentToken = [];
        const keyValuePairs = [];
        const trimmedText = text.trim();
        for (const currentChar of trimmedText) {
            const isSeparator = separator.exec(currentChar);
            if (currentChar === "'" && !inDoubleQuote) {
                inSingleQuote = !inSingleQuote;
                continue;
            }
            else if (currentChar === '"' && !inSingleQuote) {
                inDoubleQuote = !inDoubleQuote;
                continue;
            }
            if (!inSingleQuote && !inDoubleQuote && isSeparator) {
                if (currentToken.length > 0) {
                    keyValuePairs.push(currentToken.join(''));
                    currentToken = [];
                }
            }
            else {
                currentToken.push(currentChar);
            }
        }
        // For the case of only one key=value pair with no separator
        if (currentToken.length > 0) {
            keyValuePairs.push(currentToken.join(''));
        }
        return keyValuePairs;
    }
}
exports.DataCommand = DataCommand;
DataCommand.metrics = [];
