// @salesforce/plugin-data/lib/reporters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormatTypes = exports.JsonReporter = exports.CsvReporter = exports.HumanReporter = exports.QueryReporter = exports.Reporter = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const os_1 = require("os");
const core_1 = require("@salesforce/core");
const chalk = require("chalk");
const ts_types_1 = require("@salesforce/ts-types");
const dataSoqlQueryTypes_1 = require("./dataSoqlQueryTypes");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-data', 'soql.query');
class Reporter {
    constructor(ux, logger) {
        this.ux = ux;
        this.logger = logger.child('reporter');
    }
    log(...args) {
        this.ux.log(...args);
    }
}
exports.Reporter = Reporter;
class QueryReporter extends Reporter {
    constructor(data, columns, ux, logger) {
        super(ux, logger);
        this.columns = [];
        this.columns = columns;
        this.data = data;
    }
}
exports.QueryReporter = QueryReporter;
class HumanReporter extends QueryReporter {
    constructor(data, columns, ux, logger) {
        super(data, columns, ux, logger);
    }
    display() {
        const { attributeNames, children, aggregates } = this.parseFields();
        // in case of count() there are no records, but there is a totalSize
        const totalCount = this.data.result.records.length || this.data.result.totalSize;
        this.soqlQuery(attributeNames, this.massageRows(this.data.result.records, children, aggregates), totalCount);
    }
    parseFields() {
        const fields = this.columns;
        // Field names
        const attributeNames = [];
        // For subqueries. Display the children under the parents
        const children = [];
        // For function fields, like avg(total).
        const aggregates = [];
        if (fields) {
            this.logger.info(`Found fields ${JSON.stringify(fields.map((field) => `${typeof field}.${field.name}`))}`);
            fields.forEach((field) => {
                var _a;
                if (field.fieldType === dataSoqlQueryTypes_1.FieldType.subqueryField) {
                    children.push(field.name);
                    ((_a = field.fields) !== null && _a !== void 0 ? _a : []).forEach((subfield) => attributeNames.push(`${field.name}.${subfield.name}`));
                }
                else if (field.fieldType === dataSoqlQueryTypes_1.FieldType.functionField) {
                    if (field.alias) {
                        attributeNames.push(field.alias);
                    }
                    else {
                        attributeNames.push(field.name);
                    }
                    aggregates.push(field);
                }
                else {
                    attributeNames.push(field.name);
                }
            });
        }
        else {
            this.logger.info(`No fields found for query "${this.data.query}"`);
        }
        return { attributeNames, children, aggregates };
    }
    soqlQuery(columns, records, totalCount) {
        this.prepNullValues(records);
        this.ux.table(records, this.prepColumns(columns));
        this.log(chalk.bold(messages.getMessage('displayQueryRecordsRetrieved', [totalCount])));
    }
    prepNullValues(records) {
        records.forEach((record) => {
            const recordAsObject = record;
            Reflect.ownKeys(recordAsObject).forEach((propertyKey) => {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                const value = Reflect.get(recordAsObject, propertyKey);
                if (value === null) {
                    Reflect.set(recordAsObject, propertyKey, chalk.bold('null'));
                }
                else if (typeof value === 'object') {
                    this.prepNullValues([value]);
                }
            });
        });
    }
    prepColumns(columns) {
        const formattedColumns = {};
        columns
            .map((field) => field)
            .filter((field) => field)
            .map((field) => (formattedColumns[field] = {
            header: field.toUpperCase(),
            get: (row) => {
                // first test if key exists, if so, return value
                if (Reflect.has(row, field)) {
                    return Reflect.get(row, field) || '';
                }
                else {
                    // if not, try to find it query
                    return (0, ts_types_1.get)(row, field) || '';
                }
            },
        }));
        return formattedColumns;
    }
    //  public massageRows(queryResults: BasicRecord[], children: string[], aggregates: Field[]): BasicRecord[] {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    massageRows(queryResults, children, aggregates) {
        // some fields will return a JSON object that isn't accessible via the query (SELECT Metadata FROM RemoteProxy)
        // some will return a JSON that IS accessible via the query (SELECT owner.Profile.Name FROM Lead)
        // querying (SELECT Metadata.isActive FROM RemoteProxy) throws a SOQL validation error, so we have to display the entire Metadata object
        queryResults.forEach((qr) => {
            const result = qr;
            this.data.columns.forEach((col) => {
                var _a;
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                const entry = Reflect.get(result, col.name);
                if (typeof entry === 'object' && col.fieldType === dataSoqlQueryTypes_1.FieldType.field) {
                    Reflect.set(result, col.name, JSON.stringify(entry, null, 2));
                }
                else if (typeof entry === 'object' && ((_a = col.fields) === null || _a === void 0 ? void 0 : _a.length) && entry) {
                    col.fields.forEach((field) => {
                        Reflect.set(result, `${col.name}.${field.name}`, (0, ts_types_1.get)(result, `${col.name}.records[0].${field.name}`));
                    });
                }
            });
        });
        // There are subqueries or aggregates. Massage the data.
        let qr;
        if (children.length > 0 || aggregates.length > 0) {
            qr = queryResults.reduce((newResults, result) => {
                // Aggregates are soql functions that aggregate data, like "SELECT avg(total)" and
                // are returned in the data as exprX. Aggregates can have aliases, like "avg(total) totalAverage"
                // and are returned in the data as the alias.
                if (aggregates.length > 0) {
                    for (let i = 0; i < aggregates.length; i++) {
                        const aggregate = aggregates[i];
                        if (!aggregate.alias) {
                            Reflect.set(result, aggregate.name, Reflect.get(result, `expr${i}`));
                        }
                    }
                }
                const subResults = [];
                if (children.length > 0) {
                    const childrenRows = {};
                    children.forEach((child) => {
                        const aChild = (0, ts_types_1.get)(result, child);
                        Reflect.set(childrenRows, child, aChild);
                        Reflect.deleteProperty(result, child);
                    });
                    Reflect.ownKeys(childrenRows).forEach((child) => {
                        const childO = (0, ts_types_1.get)(childrenRows, child);
                        if (childO) {
                            const childRecords = (0, ts_types_1.getArray)(childO, 'records', []);
                            // eslint-disable-next-line @typescript-eslint/no-explicit-any
                            childRecords.forEach((record, index) => {
                                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                                const newResult = {};
                                Object.entries(record).forEach(([key, value]) => {
                                    if (!index) {
                                        Reflect.defineProperty(result, `${child.toString()}.${key}`, {
                                            value: value ? value : chalk.bold('null'),
                                        });
                                    }
                                    else {
                                        Reflect.defineProperty(newResult, `${child.toString()}.${key}`, {
                                            value: value ? value : chalk.bold('null'),
                                        });
                                    }
                                });
                                if (index) {
                                    subResults.push(newResult);
                                }
                            });
                        }
                    });
                }
                newResults.push(result, ...subResults);
                return newResults;
            }, []);
        }
        return qr !== null && qr !== void 0 ? qr : queryResults;
    }
}
exports.HumanReporter = HumanReporter;
const SEPARATOR = ',';
const DOUBLE_QUOTE = '"';
const SHOULD_QUOTE_REGEXP = new RegExp(`[${SEPARATOR}${DOUBLE_QUOTE}${os_1.EOL}]`);
class CsvReporter extends QueryReporter {
    constructor(data, columns, ux, logger) {
        super(data, columns, ux, logger);
    }
    /**
     * Escape a value to be placed in a CSV row. We follow rfc 4180
     * https://tools.ietf.org/html/rfc4180#section-2 and will not surround the
     * value in quotes if it doesn't contain the separator, double quote, or EOL.
     *
     * @param value The escaped value
     */
    escape(value) {
        if (value && SHOULD_QUOTE_REGEXP.test(value)) {
            return `"${value.replace(/"/gi, '""')}"`;
        }
        return value;
    }
    display() {
        const attributeNames = this.massageRows();
        // begin output
        this.log(attributeNames
            .map((name) => {
            return this.escape(name);
        })
            .join(SEPARATOR));
        this.data.result.records.forEach((row) => {
            const values = attributeNames.map((name) => {
                const value = (0, ts_types_1.get)(row, name);
                if ((0, ts_types_1.isString)(value)) {
                    return this.escape(value);
                    // if value is null, then typeof value === 'object' so check before typeof to avoid illegal csv
                }
                else if (value === null) {
                    return;
                }
                else if (typeof value === 'object') {
                    return this.escape(JSON.stringify(value));
                }
                return value;
            });
            this.log(values.join(SEPARATOR));
        });
    }
    massageRows() {
        const fields = this.columns;
        const hasSubqueries = fields.some((field) => field.fieldType === dataSoqlQueryTypes_1.FieldType.subqueryField);
        const hasFunctions = fields.some((field) => field.fieldType === dataSoqlQueryTypes_1.FieldType.functionField);
        const attributeNames = [];
        if (fields) {
            this.logger.info(`Found fields ${JSON.stringify(fields.map((field) => `${typeof field}.${field.name}`))}`);
        }
        else {
            this.logger.info(`No fields found for query "${this.data.query}"`);
        }
        if (hasSubqueries || hasFunctions) {
            // If there are subqueries, we need to get the max child length for each subquery.
            const typeLengths = new Map();
            // For function fields, like avg(total).
            const aggregates = [];
            fields.forEach((field) => {
                if (field.fieldType === dataSoqlQueryTypes_1.FieldType.subqueryField) {
                    typeLengths.set(field.name, 0);
                }
                if (field.fieldType === dataSoqlQueryTypes_1.FieldType.functionField) {
                    aggregates.push(field);
                }
            });
            // Get max lengths by iterating over the records once
            this.data.result.records.forEach((result) => {
                [...typeLengths.keys()].forEach((key) => {
                    var _a;
                    const record = (0, ts_types_1.get)(result, key);
                    const totalSize = (0, ts_types_1.getNumber)(record, 'totalSize');
                    if (!!totalSize && totalSize > ((_a = typeLengths.get(key)) !== null && _a !== void 0 ? _a : 0)) {
                        typeLengths.set(key, totalSize);
                    }
                });
                // Aggregates are soql functions that aggregate data, like "SELECT avg(total)" and
                // are returned in the data as exprX. Aggregates can have aliases, like "avg(total) totalAverage"
                // and are returned in the data as the alias.
                if (aggregates.length > 0) {
                    for (let i = 0; i < aggregates.length; i++) {
                        const aggregate = aggregates[i];
                        if (!aggregate.alias) {
                            Reflect.set(result, aggregate.name, Reflect.get(result, `expr${i}`));
                        }
                    }
                }
            });
            fields.forEach((field) => {
                var _a, _b;
                if (typeLengths.get(field.name)) {
                    for (let i = 0; i < ((_a = typeLengths.get(field.name)) !== null && _a !== void 0 ? _a : 0); i++) {
                        attributeNames.push(`${field.name}.totalSize`);
                        ((_b = field.fields) !== null && _b !== void 0 ? _b : []).forEach((subfield) => {
                            attributeNames.push(`${field.name}.records.${i}.${subfield.name}`);
                        });
                    }
                }
                else if (field.fieldType === dataSoqlQueryTypes_1.FieldType.functionField) {
                    if (field.alias) {
                        attributeNames.push(field.alias);
                    }
                    else {
                        attributeNames.push(field.name);
                    }
                }
                else {
                    attributeNames.push(field.name);
                }
            });
        }
        else {
            attributeNames.push(...fields.map((field) => field.name));
        }
        return attributeNames;
    }
}
exports.CsvReporter = CsvReporter;
class JsonReporter extends QueryReporter {
    constructor(data, columns, ux, logger) {
        super(data, columns, ux, logger);
    }
    log() {
        return;
    }
    display() {
        this.ux.styledJSON({ status: 0, result: this.data.result });
    }
}
exports.JsonReporter = JsonReporter;
/**
 * A list of the accepted reporter types
 */
exports.FormatTypes = {
    human: HumanReporter,
    csv: CsvReporter,
    json: JsonReporter,
};
