// @salesforce/plugin-source/lib/formatters/deployResultFormatter.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployResultFormatter = void 0;
const chalk = require("chalk");
const core_1 = require("@salesforce/core");
const ts_types_1 = require("@salesforce/ts-types");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
const coverageUtils_1 = require("../coverageUtils");
const resultFormatter_1 = require("./resultFormatter");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-source', 'deploy');
class DeployResultFormatter extends resultFormatter_1.ResultFormatter {
    constructor(logger, ux, options, result) {
        super(logger, ux, options);
        this.result = result;
        this.fileResponses = (result === null || result === void 0 ? void 0 : result.getFileResponses) ? result.getFileResponses() : [];
    }
    /**
     * Get the JSON output from the DeployResult.
     *
     * @returns a JSON formatted result matching the provided type.
     */
    getJson() {
        const json = this.getResponse();
        json.deployedSource = this.fileResponses;
        json.outboundFiles = []; // to match toolbelt version
        json.deploys = [Object.assign({}, this.getResponse())]; // to match toolbelt version
        if (this.options.testsRan) {
            json.coverage = this.getCoverageFileInfo();
            json.junit = this.getJunitFileInfo();
        }
        return json;
    }
    /**
     * Displays deploy results in human readable format.  Output can vary based on:
     *
     * 1. Verbose option
     * 3. Checkonly deploy (checkonly=true)
     * 4. Deploy with test results
     * 5. Canceled status
     */
    display() {
        // Display check-only, non-verbose success
        if (this.isSuccess() && this.isCheckOnly() && !this.isVerbose()) {
            const componentsDeployed = this.getNumResult('numberComponentsDeployed');
            const testsCompleted = this.getNumResult('numberTestsCompleted');
            const checkOnlySuccessMsg = messages.getMessage('checkOnlySuccess', [componentsDeployed, testsCompleted]);
            this.ux.log(checkOnlySuccessMsg);
            return;
        }
        if (this.hasStatus(source_deploy_retrieve_1.RequestStatus.Canceled)) {
            const canceledByName = (0, ts_types_1.getString)(this.result, 'response.canceledByName', 'unknown');
            throw new core_1.SfError(messages.getMessage('deployCanceled', [canceledByName]), 'DeployFailed');
        }
        this.displaySuccesses();
        this.displayDeletions();
        this.displayFailures();
        this.displayTestResults();
        this.displayOutputFileLocations();
        // Throw a DeployFailed error unless the deployment was successful.
        if (!this.isSuccess()) {
            throw new core_1.SfError(messages.getMessage('deployFailed'), 'DeployFailed');
        }
        this.ux.log(messages.getMessage(this.isCheckOnly() ? 'checkOnlySuccessVerbose' : 'deploySuccess'));
    }
    hasStatus(status) {
        return (0, ts_types_1.getString)(this.result, 'response.status') === status;
    }
    isRunTestsEnabled() {
        return (0, ts_types_1.getBoolean)(this.result, 'response.runTestsEnabled', false);
    }
    isCheckOnly() {
        return (0, ts_types_1.getBoolean)(this.result, 'response.checkOnly', false);
    }
    getNumResult(field) {
        return (0, ts_types_1.getNumber)(this.result, `response.${field}`, 0);
    }
    getResponse() {
        return (0, ts_types_1.get)(this.result, 'response', {});
    }
    displaySuccesses() {
        var _a;
        if (this.isSuccess() && ((_a = this.fileResponses) === null || _a === void 0 ? void 0 : _a.length)) {
            const successes = this.fileResponses.filter((f) => !['Failed', 'Deleted'].includes(f.state));
            if (!successes.length) {
                return;
            }
            this.sortFileResponses(successes);
            this.asRelativePaths(successes);
            this.ux.log('');
            this.ux.styledHeader(chalk.blue('Deployed Source'));
            this.ux.table(successes, {
                fullName: { header: 'FULL NAME' },
                type: { header: 'TYPE' },
                filePath: { header: 'PROJECT PATH' },
            });
        }
    }
    displayDeletions() {
        const deletions = this.fileResponses.filter((f) => f.state === 'Deleted');
        if (!deletions.length) {
            return;
        }
        this.sortFileResponses(deletions);
        this.asRelativePaths(deletions);
        this.ux.log('');
        this.ux.styledHeader(chalk.blue('Deleted Source'));
        this.ux.table(deletions, {
            fullName: { header: 'FULL NAME' },
            type: { header: 'TYPE' },
            filePath: { header: 'PROJECT PATH' },
        });
    }
    displayFailures() {
        var _a, _b, _c, _d;
        if (this.hasStatus(source_deploy_retrieve_1.RequestStatus.Failed) || this.hasStatus(source_deploy_retrieve_1.RequestStatus.SucceededPartial)) {
            const failures = [];
            const fileResponseFailures = new Map();
            if ((_a = this.fileResponses) === null || _a === void 0 ? void 0 : _a.length) {
                const fileResponses = [];
                this.fileResponses
                    .filter((f) => f.state === 'Failed')
                    .map((f) => {
                    fileResponses.push(f);
                    fileResponseFailures.set(`${f.type}#${f.fullName}`, f.error);
                });
                this.sortFileResponses(fileResponses);
                this.asRelativePaths(fileResponses);
                failures.push(...fileResponses);
            }
            const deployMessages = (0, resultFormatter_1.toArray)((_d = (_c = (_b = this.result) === null || _b === void 0 ? void 0 : _b.response) === null || _c === void 0 ? void 0 : _c.details) === null || _d === void 0 ? void 0 : _d.componentFailures);
            if (deployMessages.length > failures.length) {
                // if there's additional failures in the API response, find the failure and add it to the output
                deployMessages.map((deployMessage) => {
                    if (!fileResponseFailures.has(`${deployMessage.componentType}#${deployMessage.fullName}`)) {
                        // duplicate the problem message to the error property for displaying in the table
                        failures.push(Object.assign(deployMessage, { error: deployMessage.problem }));
                    }
                });
            }
            this.ux.log('');
            this.ux.styledHeader(chalk.red(`Component Failures [${failures.length}]`));
            this.ux.table(failures, {
                problemType: { header: 'Type' },
                fullName: { header: 'Name' },
                error: { header: 'Problem' },
            });
            this.ux.log('');
        }
    }
    displayTestResults() {
        if (this.isRunTestsEnabled()) {
            this.ux.log('');
            if (this.isVerbose()) {
                this.verboseTestFailures();
                this.verboseTestSuccess();
                this.verboseTestTime();
            }
            else {
                this.ux.styledHeader(chalk.blue('Test Results Summary'));
                this.ux.log(`Passing: ${this.getNumResult('numberTestsCompleted')}`);
                this.ux.log(`Failing: ${this.getNumResult('numberTestErrors')}`);
                this.ux.log(`Total: ${this.getNumResult('numberTestsTotal')}`);
                this.ux.log(`Time: ${this.getNumResult('details.runTestResult.totalTime')}`);
            }
        }
    }
    verboseTestFailures() {
        var _a, _b, _c, _d, _e;
        if ((_b = (_a = this.result) === null || _a === void 0 ? void 0 : _a.response) === null || _b === void 0 ? void 0 : _b.numberTestErrors) {
            const failures = (0, resultFormatter_1.toArray)((_d = (_c = this.result.response.details) === null || _c === void 0 ? void 0 : _c.runTestResult) === null || _d === void 0 ? void 0 : _d.failures);
            const tests = this.sortTestResults(failures);
            this.ux.log('');
            this.ux.styledHeader(chalk.red(`Test Failures [${(0, ts_types_1.asString)((_e = this.result.response.details.runTestResult) === null || _e === void 0 ? void 0 : _e.numFailures)}]`));
            this.ux.table(tests, {
                name: { header: 'Name' },
                methodName: { header: 'Method' },
                message: { header: 'Message' },
                stackTrace: { header: 'Stacktrace' },
            });
        }
    }
    verboseTestSuccess() {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const success = (0, resultFormatter_1.toArray)((_d = (_c = (_b = (_a = this.result) === null || _a === void 0 ? void 0 : _a.response) === null || _b === void 0 ? void 0 : _b.details) === null || _c === void 0 ? void 0 : _c.runTestResult) === null || _d === void 0 ? void 0 : _d.successes);
        if (success.length) {
            const tests = this.sortTestResults(success);
            this.ux.log('');
            this.ux.styledHeader(chalk.green(`Test Success [${success.length}]`));
            this.ux.table(tests, {
                name: { header: 'Name' },
                methodName: { header: 'Method' },
            });
        }
        const codeCoverage = (0, resultFormatter_1.toArray)((_h = (_g = (_f = (_e = this.result) === null || _e === void 0 ? void 0 : _e.response) === null || _f === void 0 ? void 0 : _f.details) === null || _g === void 0 ? void 0 : _g.runTestResult) === null || _h === void 0 ? void 0 : _h.codeCoverage);
        if (codeCoverage.length) {
            const coverage = (0, coverageUtils_1.prepCoverageForDisplay)(codeCoverage);
            this.ux.log('');
            this.ux.styledHeader(chalk.blue('Apex Code Coverage'));
            this.ux.table(coverage, {
                name: { header: 'Name' },
                numLocations: { header: '% Covered' },
                lineNotCovered: { header: 'Uncovered Lines' },
            });
        }
    }
    verboseTestTime() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        if (((_c = (_b = (_a = this.result.response) === null || _a === void 0 ? void 0 : _a.details) === null || _b === void 0 ? void 0 : _b.runTestResult) === null || _c === void 0 ? void 0 : _c.successes) ||
            ((_g = (_f = (_e = (_d = this.result) === null || _d === void 0 ? void 0 : _d.response) === null || _e === void 0 ? void 0 : _e.details) === null || _f === void 0 ? void 0 : _f.runTestResult) === null || _g === void 0 ? void 0 : _g.failures)) {
            this.ux.log('');
            this.ux.log(`Total Test Time:  ${(_l = (_k = (_j = (_h = this.result) === null || _h === void 0 ? void 0 : _h.response) === null || _j === void 0 ? void 0 : _j.details) === null || _k === void 0 ? void 0 : _k.runTestResult) === null || _l === void 0 ? void 0 : _l.totalTime}`);
        }
    }
}
exports.DeployResultFormatter = DeployResultFormatter;
