// @salesforce/plugin-org/lib/shared/orgHighlighter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyledObject = exports.getStyledValue = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const chalk = require("chalk");
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
const styledProperties = {
    status: {
        Active: 'green',
        else: 'red',
    },
    connectedStatus: {
        Connected: 'green',
        else: 'red',
    },
};
const getStyledValue = (key, value) => {
    if (styledProperties[key] && value) {
        return styledProperties[key][value]
            ? chalk[styledProperties[key][value]](value)
            : chalk[styledProperties[key].else](value);
    }
    return value;
};
exports.getStyledValue = getStyledValue;
const getStyledObject = (objectToStyle) => {
    const clonedObject = { ...objectToStyle };
    for (const key of Object.keys(styledProperties)) {
        if (Reflect.has(clonedObject, key)) {
            clonedObject[key] = (0, exports.getStyledValue)(key, clonedObject[key]);
        }
    }
    return clonedObject;
};
exports.getStyledObject = getStyledObject;
