// salesforce-alm/dist/commands/force/package1/version/list.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Package1VersionListCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
class Package1VersionListCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const Package1VersionListCommandImpl = require('../../../../lib/package1/package1VersionListCommand');
        return this.execLegacyCommand(new Package1VersionListCommandImpl(context.org), context);
    }
}
exports.Package1VersionListCommand = Package1VersionListCommand;
Package1VersionListCommand.theDescription = messages.getMessage('package1VersionListCommandCliDescription');
Package1VersionListCommand.longDescription = messages.getMessage('package1VersionListCommandLongDescription');
Package1VersionListCommand.help = messages.getMessage('package1VersionListCommandCliHelp');
Package1VersionListCommand.supportsUsername = true;
Package1VersionListCommand.requiresProject = true;
Package1VersionListCommand.flagsConfig = {
    packageid: command_1.flags.id({
        char: 'i',
        description: messages.getMessage('package1VersionListCommandPackageId'),
        longDescription: messages.getMessage('package1VersionListCommandPackageIdLong'),
        required: false,
    }),
};
