// @salesforce/plugin-community/lib/shared/community/connect/CommunityCreateResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommunityCreateResource = void 0;
const core_1 = require("@salesforce/core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-community', 'create');
const MESSAGE_KEY = 'message';
const NAME_KEY = 'name';
const ACTION_KEY = 'action';
/**
 * A connect api resource for creating a community
 */
class CommunityCreateResource {
    constructor(flags, params, ux) {
        this.flags = flags;
        this.params = params;
        this.ux = ux;
    }
    handleSuccess(result) {
        const response = {
            message: messages.getMessage('response.createMessage'),
            name: result[NAME_KEY],
            action: messages.getMessage('response.action'),
        };
        const columns = {
            [NAME_KEY]: { header: 'Name' },
            [MESSAGE_KEY]: { header: 'Message' },
            [ACTION_KEY]: { header: 'Action' },
        };
        this.ux.styledHeader(messages.getMessage('response.styleHeader'));
        this.ux.table([response], columns);
        return response;
    }
    handleError(error) {
        throw error;
    }
    // eslint-disable-next-line @typescript-eslint/require-await
    async fetchRelativeConnectUrl() {
        return '/connect/communities';
    }
    getRequestMethod() {
        return 'POST';
    }
    // eslint-disable-next-line @typescript-eslint/require-await
    async fetchPostParams() {
        const params = {
            name: this.flags.name,
            urlPathPrefix: this.flags.urlpathprefix,
            templateName: this.flags.templatename,
            description: this.flags.description,
            templateParams: this.params['templateParams'],
        };
        return JSON.stringify(params);
    }
}
exports.CommunityCreateResource = CommunityCreateResource;
