// @salesforce/plugin-source/lib/formatters/source/deployAsyncResultFormatter.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployAsyncResultFormatter = void 0;
const os_1 = require("os");
const core_1 = require("@salesforce/core");
const kit_1 = require("@salesforce/kit");
const resultFormatter_1 = require("../resultFormatter");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-source', 'deploy');
class DeployAsyncResultFormatter extends resultFormatter_1.ResultFormatter {
    constructor(logger, ux, options, result) {
        super(logger, ux, options);
        this.result = result;
    }
    /**
     * Get the JSON output from the AsyncDeployResult.
     *
     * @returns a DeployCommandAsyncResult.
     */
    getJson() {
        const asyncResult = {
            id: this.result.id,
            done: false,
            state: 'Queued',
            status: 'Queued',
            timedOut: true,
        };
        const resultClone = (0, kit_1.cloneJson)(asyncResult);
        asyncResult.outboundFiles = [];
        asyncResult.deploys = [resultClone];
        return asyncResult;
    }
    /**
     * Displays async deploy results in human format.
     */
    display() {
        this.ux.log(messages.getMessage('asyncDeployQueued'), os_1.EOL);
        this.ux.log(messages.getMessage('asyncDeployCancel', [this.result.id, this.options.username]));
        this.ux.log(messages.getMessage('asyncDeployReport', [this.result.id, this.options.username]));
    }
}
exports.DeployAsyncResultFormatter = DeployAsyncResultFormatter;
