// @salesforce/plugin-source/lib/formatters/source/pullFormatter.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PullResultFormatter = void 0;
const chalk_1 = require("chalk");
const core_1 = require("@salesforce/core");
const ts_types_1 = require("@salesforce/ts-types");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
const resultFormatter_1 = require("../resultFormatter");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.load('@salesforce/plugin-source', 'pull', [
    'retrievedSourceWarningsHeader',
    'retrievedSourceHeader',
    'retrieveTimeout',
    'NoResultsFound',
]);
class PullResultFormatter extends resultFormatter_1.ResultFormatter {
    constructor(logger, ux, options, retrieveResult, deleteResult = []) {
        var _a, _b, _c, _d;
        super(logger, ux, options);
        this.result = retrieveResult;
        this.fileResponses = ((retrieveResult === null || retrieveResult === void 0 ? void 0 : retrieveResult.getFileResponses) ? retrieveResult.getFileResponses() : []).concat(deleteResult);
        const warnMessages = (_b = (_a = retrieveResult === null || retrieveResult === void 0 ? void 0 : retrieveResult.response) === null || _a === void 0 ? void 0 : _a.messages) !== null && _b !== void 0 ? _b : [];
        this.warnings = (0, resultFormatter_1.toArray)(warnMessages);
        if ((_d = (_c = this.result) === null || _c === void 0 ? void 0 : _c.response) === null || _d === void 0 ? void 0 : _d.zipFile) {
            // zipFile can become massive and unwieldy with JSON parsing/terminal output and, isn't useful
            delete this.result.response.zipFile;
        }
    }
    /**
     * Get the JSON output from the PullCommandResult.
     *
     * @returns RetrieveCommandResult
     */
    getJson() {
        if (!this.isSuccess()) {
            const error = new core_1.SfError('Pull failed.', 'PullFailed', [], process.exitCode);
            error.setData(this.fileResponses.map(({ state, fullName, type, filePath }) => ({ state, fullName, type, filePath })));
            throw error;
        }
        return {
            pulledSource: this.fileResponses.map(({ state, fullName, type, filePath }) => ({
                state,
                fullName,
                type,
                filePath,
            })),
        };
    }
    /**
     * Displays retrieve results in human format.
     */
    display() {
        if (this.hasStatus(source_deploy_retrieve_1.RequestStatus.InProgress)) {
            const commandWaitTime = (0, ts_types_1.getNumber)(this.options, 'waitTime', 33);
            this.ux.log(messages.getMessage('retrieveTimeout', [commandWaitTime]));
            return;
        }
        if (this.isSuccess()) {
            this.ux.styledHeader((0, chalk_1.blue)(messages.getMessage('retrievedSourceHeader')));
            const retrievedFiles = this.fileResponses.filter((fr) => fr.state !== source_deploy_retrieve_1.ComponentStatus.Failed);
            if (retrievedFiles === null || retrievedFiles === void 0 ? void 0 : retrievedFiles.length) {
                this.displaySuccesses(retrievedFiles);
            }
            else {
                this.ux.log(messages.getMessage('NoResultsFound'));
            }
            if (this.warnings.length) {
                this.displayWarnings();
            }
        }
        else {
            this.displayErrors();
        }
    }
    hasStatus(status) {
        return (0, ts_types_1.getString)(this.result, 'response.status') === status;
    }
    hasComponents() {
        return (0, ts_types_1.getNumber)(this.result, 'components.size', 0) === 0;
    }
    displayWarnings() {
        this.ux.styledHeader((0, chalk_1.yellow)(messages.getMessage('retrievedSourceWarningsHeader')));
        this.ux.table(this.warnings, { fileName: { header: 'FILE NAME' }, problem: { header: 'PROBLEM' } });
        this.ux.log();
    }
    displaySuccesses(retrievedFiles) {
        this.sortFileResponses(retrievedFiles);
        this.asRelativePaths(retrievedFiles);
        this.ux.table(retrievedFiles, {
            state: { header: 'STATE' },
            fullName: { header: 'FULL NAME' },
            type: { header: 'TYPE' },
            filePath: { header: 'PROJECT PATH' },
        });
    }
    displayErrors() {
        // an invalid packagename retrieval will end up with a message in the `errorMessage` entry
        if (!this.result) {
            return;
        }
        const errorMessage = (0, ts_types_1.getString)(this.result.response, 'errorMessage');
        if (errorMessage) {
            throw new core_1.SfError(errorMessage);
        }
        const unknownMsg = [{ fileName: 'unknown', problem: 'unknown' }];
        const responseMsgs = (0, ts_types_1.get)(this.result, 'response.messages', unknownMsg);
        const errMsgs = (0, resultFormatter_1.toArray)(responseMsgs);
        const errMsgsForDisplay = errMsgs.reduce((p, c) => `${p}\n${c.fileName}: ${c.problem}`, '');
        this.ux.log(`Retrieve Failed due to: ${errMsgsForDisplay}`);
    }
}
exports.PullResultFormatter = PullResultFormatter;
