// salesforce-alm/dist/commands/force/package1/version/create/get.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Package1VersionCreateGetCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../../ToolbeltCommand");
const messages = Messages();
class Package1VersionCreateGetCommand extends ToolbeltCommand_1.ToolbeltCommand {
    /**
     * context hook
     *
     * @param context - the cli context
     * @returns {Promise}
     */
    async run() {
        const context = await this.resolveLegacyContext();
        const Package1VersionCreateGetCommandImpl = require('../../../../../lib/package1/package1VersionCreateGetCommand');
        return this.execLegacyCommand(new Package1VersionCreateGetCommandImpl(), context);
    }
}
exports.Package1VersionCreateGetCommand = Package1VersionCreateGetCommand;
Package1VersionCreateGetCommand.theDescription = messages.getMessage('cliDescription', [], 'package1_version_create_get');
Package1VersionCreateGetCommand.longDescription = messages.getMessage('cliDescriptionLong', [], 'package1_version_create_get');
Package1VersionCreateGetCommand.help = messages.getMessage('help', [], 'package1_version_create_get');
Package1VersionCreateGetCommand.requiresProject = false;
Package1VersionCreateGetCommand.requiresUsername = true;
Package1VersionCreateGetCommand.flagsConfig = {
    requestid: command_1.flags.id({
        char: 'i',
        description: messages.getMessage('requestId', [], 'package1_version_create_get'),
        longDescription: messages.getMessage('requestIdLong', [], 'package1_version_create_get'),
        required: true,
    }),
};
