// @salesforce/plugin-templates/lib/commands/force/project/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const projectGenerator_1 = require("@salesforce/templates/lib/generators/projectGenerator");
const utils_1 = require("../../../utils");
class Project extends utils_1.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            // namespace is a reserved keyword for the generator
            this.flags.ns = this.flags.namespace;
            return this.runGenerator(projectGenerator_1.default);
        });
    }
}
exports.default = Project;
Project.description = utils_1.MessageUtil.buildDescription('ProjectDescription', false);
Project.examples = [
    '$ sfdx force:project:create --projectname mywork',
    '$ sfdx force:project:create --projectname mywork --defaultpackagedir myapp',
    '$ sfdx force:project:create --projectname mywork --defaultpackagedir myapp --manifest',
    '$ sfdx force:project:create --projectname mywork --template empty'
];
Project.help = utils_1.MessageUtil.buildHelpText(Project.examples, false);
Project.longDescription = utils_1.MessageUtil.get('ProjectLongDescription');
Project.flagsConfig = {
    projectname: command_1.flags.string({
        char: 'n',
        description: utils_1.MessageUtil.get('ProjectNameFlagDescription'),
        longDescription: utils_1.MessageUtil.get('ProjectNameFlagLongDescription'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: utils_1.MessageUtil.get('ProjectTemplateFlagDescription'),
        longDescription: utils_1.MessageUtil.get('ProjectTemplateFlagLongDescription'),
        default: 'standard',
        options: ['standard', 'empty', 'analytics']
    }),
    outputdir: command_1.flags.string({
        char: 'd',
        description: utils_1.MessageUtil.get('OutputDirFlagDescription'),
        longDescription: utils_1.MessageUtil.get('OutputDirFlagLongDescription'),
        default: '.'
    }),
    namespace: command_1.flags.string({
        char: 's',
        description: utils_1.MessageUtil.get('ProjectNamespaceFlagDescription'),
        longDescription: utils_1.MessageUtil.get('ProjectNamespaceFlagLongDescription'),
        default: ''
    }),
    defaultpackagedir: command_1.flags.string({
        char: 'p',
        description: utils_1.MessageUtil.get('ProjectPackageFlagDescription'),
        longDescription: utils_1.MessageUtil.get('ProjectPackageFlagLongDescription'),
        default: 'force-app'
    }),
    manifest: command_1.flags.boolean({
        char: 'x',
        description: utils_1.MessageUtil.get('ProjectManifestFlagDescription'),
        longDescription: utils_1.MessageUtil.get('ProjectManifestFlagLongDescription')
    }),
    loginurl: command_1.flags.string({
        char: 'l',
        description: utils_1.MessageUtil.get('ProjectLoginUrlDescription'),
        longDescription: utils_1.MessageUtil.get('ProjectLoginUrlLongDescription'),
        default: 'https://login.salesforce.com',
        hidden: true
    })
};
