// salesforce-alm/dist/lib/source/decompositionStrategy/describeMetadataDecompositionConfig.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DescribeMetadataDecompositionConfig = exports.DescribeMetadataDecomposedSubtypeConfig = void 0;
/**
 * Subtype configuration for the describeMetadata decomposition.
 */
class DescribeMetadataDecomposedSubtypeConfig {
}
exports.DescribeMetadataDecomposedSubtypeConfig = DescribeMetadataDecomposedSubtypeConfig;
/**
 * Decomposition configuration for the describeMetadata decomposition and the folder per subtype
 * workspace manifestation.
 */
class DescribeMetadataDecompositionConfig {
    constructor(metadataName, isGlobal, isEmptyContainer, useSparseComposition) {
        this.strategy = 'describeMetadata';
        this.workspaceStrategy = 'folderPerSubtype';
        this.commitStrategy = 'fineGrainTracking';
        this.contentStrategy = 'N/A';
        this.metadataName = metadataName;
        this.isGlobal = isGlobal;
        this.isEmptyContainer = isEmptyContainer;
        this.useSparseComposition = useSparseComposition;
        this.decompositions = [];
    }
}
exports.DescribeMetadataDecompositionConfig = DescribeMetadataDecompositionConfig;
