// salesforce-alm/dist/lib/source/decompositionStrategy/inFolderDecompositionConfig.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InFolderDecompositionConfig = void 0;
/**
 * Decomposition config for Documents, Reports, Email Templates, and Dashboards
 */
class InFolderDecompositionConfig {
    constructor(metadataName, isGlobal, isEmptyContainer) {
        this.strategy = 'describeMetadata';
        this.workspaceStrategy = 'inFolderMetadataType';
        this.commitStrategy = 'fineGrainTracking';
        this.contentStrategy = 'nonDecomposedContent';
        this.metadataName = metadataName;
        this.isGlobal = isGlobal;
        this.isEmptyContainer = isEmptyContainer;
        this.useSparseComposition = false;
        this.decompositions = [];
    }
}
exports.InFolderDecompositionConfig = InFolderDecompositionConfig;
