// salesforce-alm/dist/lib/source/decompositionStrategy/inFolderMetadataWorkspaceDecomposition.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint-disable @typescript-eslint/no-unused-vars */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InFolderMetadataWorkspaceDecomposition = void 0;
class InFolderMetadataWorkspaceDecomposition {
    getDecomposedFileName(annotation, decomposedSubtypeConfig) {
        return null;
    }
    getContainerPath(metadataFilePath, ext) {
        return metadataFilePath;
    }
    findDecomposedPaths(metadataFilePath, ext) {
        return new Map();
    }
    getDecomposedSubtypeDirFromMetadataFile(metadataFilePath, ext, decomposedSubtypeConfig) {
        return null;
    }
    getDecomposedSubtypeDirFromAnnotation(annotation, metadataType, aggregateFullName, decomposedSubtypeConfig) {
        return null;
    }
}
exports.InFolderMetadataWorkspaceDecomposition = InFolderMetadataWorkspaceDecomposition;
