// salesforce-alm/dist/commands/force/package/install/report.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageInstallReportCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../../lib/messages");
const ToolbeltCommand_1 = require("../../../../ToolbeltCommand");
const messages = Messages();
class PackageInstallReportCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const PackageInstallRequestReportCommand = require('../../../../lib/package/PackageInstallRequestReportCommand');
        return this.execLegacyCommand(new PackageInstallRequestReportCommand(), context);
    }
}
exports.PackageInstallReportCommand = PackageInstallReportCommand;
PackageInstallReportCommand.theDescription = messages.getMessage('cliDescription', [], 'package_install_report');
PackageInstallReportCommand.longDescription = messages.getMessage('cliDescriptionLong', [], 'package_install_report');
PackageInstallReportCommand.help = messages.getMessage('help', [], 'package_install_report');
PackageInstallReportCommand.requiresProject = false;
PackageInstallReportCommand.requiresUsername = true;
PackageInstallReportCommand.flagsConfig = {
    requestid: command_1.flags.id({
        char: 'i',
        description: messages.getMessage('requestId', [], 'package_install_report'),
        longDescription: messages.getMessage('requestIdLong', [], 'package_install_report'),
        required: true,
    }),
};
