// @salesforce/source-deploy-retrieve/lib/src/resolve/adapters/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseSourceAdapter = exports.DefaultSourceAdapter = exports.DecomposedSourceAdapter = exports.MixedContentSourceAdapter = exports.BundleSourceAdapter = exports.MatchingContentSourceAdapter = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var matchingContentSourceAdapter_1 = require("./matchingContentSourceAdapter");
Object.defineProperty(exports, "MatchingContentSourceAdapter", { enumerable: true, get: function () { return matchingContentSourceAdapter_1.MatchingContentSourceAdapter; } });
var bundleSourceAdapter_1 = require("./bundleSourceAdapter");
Object.defineProperty(exports, "BundleSourceAdapter", { enumerable: true, get: function () { return bundleSourceAdapter_1.BundleSourceAdapter; } });
var mixedContentSourceAdapter_1 = require("./mixedContentSourceAdapter");
Object.defineProperty(exports, "MixedContentSourceAdapter", { enumerable: true, get: function () { return mixedContentSourceAdapter_1.MixedContentSourceAdapter; } });
var decomposedSourceAdapter_1 = require("./decomposedSourceAdapter");
Object.defineProperty(exports, "DecomposedSourceAdapter", { enumerable: true, get: function () { return decomposedSourceAdapter_1.DecomposedSourceAdapter; } });
var defaultSourceAdapter_1 = require("./defaultSourceAdapter");
Object.defineProperty(exports, "DefaultSourceAdapter", { enumerable: true, get: function () { return defaultSourceAdapter_1.DefaultSourceAdapter; } });
var baseSourceAdapter_1 = require("./baseSourceAdapter");
Object.defineProperty(exports, "BaseSourceAdapter", { enumerable: true, get: function () { return baseSourceAdapter_1.BaseSourceAdapter; } });
