// @salesforce/source-deploy-retrieve/lib/src/resolve/adapters/sourceAdapterFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceAdapterFactory = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const core_1 = require("@salesforce/core");
const forceIgnore_1 = require("../forceIgnore");
const bundleSourceAdapter_1 = require("./bundleSourceAdapter");
const decomposedSourceAdapter_1 = require("./decomposedSourceAdapter");
const matchingContentSourceAdapter_1 = require("./matchingContentSourceAdapter");
const mixedContentSourceAdapter_1 = require("./mixedContentSourceAdapter");
const defaultSourceAdapter_1 = require("./defaultSourceAdapter");
var AdapterId;
(function (AdapterId) {
    AdapterId["Bundle"] = "bundle";
    AdapterId["Decomposed"] = "decomposed";
    AdapterId["Default"] = "default";
    AdapterId["MatchingContentFile"] = "matchingContentFile";
    AdapterId["MixedContent"] = "mixedContent";
})(AdapterId || (AdapterId = {}));
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.load('@salesforce/source-deploy-retrieve', 'sdr', ['error_missing_adapter']);
class SourceAdapterFactory {
    constructor(registry, tree) {
        this.registry = registry;
        this.tree = tree;
    }
    getAdapter(type, forceIgnore = new forceIgnore_1.ForceIgnore()) {
        const adapterId = type.strategies?.adapter;
        switch (adapterId) {
            case AdapterId.Bundle:
                return new bundleSourceAdapter_1.BundleSourceAdapter(type, this.registry, forceIgnore, this.tree);
            case AdapterId.Decomposed:
                return new decomposedSourceAdapter_1.DecomposedSourceAdapter(type, this.registry, forceIgnore, this.tree);
            case AdapterId.MatchingContentFile:
                return new matchingContentSourceAdapter_1.MatchingContentSourceAdapter(type, this.registry, forceIgnore, this.tree);
            case AdapterId.MixedContent:
                return new mixedContentSourceAdapter_1.MixedContentSourceAdapter(type, this.registry, forceIgnore, this.tree);
            case AdapterId.Default:
                return new defaultSourceAdapter_1.DefaultSourceAdapter(type, this.registry, forceIgnore, this.tree);
            case undefined:
                return new defaultSourceAdapter_1.DefaultSourceAdapter(type, this.registry, forceIgnore, this.tree);
            default:
                throw new core_1.SfError(messages.getMessage('error_missing_adapter', [adapterId, type.name]), 'RegistryError');
        }
    }
}
exports.SourceAdapterFactory = SourceAdapterFactory;
