// salesforce-alm/dist/commands/force/package/delete.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageDeleteCommand = void 0;
// This is the legacy converted command file. Ignoring code-coverage since this is generated.
// THIS SHOULD BE REMOVED WHEN CONVERTED TO EXTEND SfdxCommand
/* istanbul ignore file */
const command_1 = require("@salesforce/command");
const Messages = require("../../../lib/messages");
const ToolbeltCommand_1 = require("../../../ToolbeltCommand");
const messages = Messages();
const consts = require("../../../lib/core/constants");
class PackageDeleteCommand extends ToolbeltCommand_1.ToolbeltCommand {
    async run() {
        const context = await this.resolveLegacyContext();
        const PackageDeleteCommandImpl = require('../../../lib/package/packageDeleteCommand');
        const heroku = require('heroku-cli-util');
        return this.execLegacyCommand(new PackageDeleteCommandImpl(heroku.prompt), context);
    }
}
exports.PackageDeleteCommand = PackageDeleteCommand;
PackageDeleteCommand.theDescription = messages.getMessage('cliDescription', [], 'package_delete');
PackageDeleteCommand.longDescription = messages.getMessage('cliLongDescription', [], 'package_delete');
PackageDeleteCommand.help = messages.getMessage('help', [], 'package_delete');
PackageDeleteCommand.orgType = consts.DEFAULT_DEV_HUB_USERNAME;
PackageDeleteCommand.requiresDevhubUsername = true;
PackageDeleteCommand.requiresProject = true;
PackageDeleteCommand.flagsConfig = {
    noprompt: command_1.flags.boolean({
        char: 'n',
        description: messages.getMessage('noPrompt', [], 'package_delete'),
        longDescription: messages.getMessage('noPrompt', [], 'package_delete'),
        required: false,
    }),
    package: command_1.flags.string({
        char: 'p',
        description: messages.getMessage('package', [], 'package_delete'),
        longDescription: messages.getMessage('packageLong', [], 'package_delete'),
        required: true,
    }),
    undelete: command_1.flags.boolean({
        description: messages.getMessage('undelete', [], 'package_delete'),
        longDescription: messages.getMessage('undeleteLong', [], 'package_delete'),
        required: false,
        hidden: true,
    }),
};
