// @salesforce/source-deploy-retrieve/lib/src/registry/coverage.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
/* eslint-disable no-console */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCoverage = exports.getCurrentApiVersion = void 0;
const got_1 = require("got");
const ProxyAgent = require("proxy-agent");
const proxy_from_env_1 = require("proxy-from-env");
const getProxiedOptions = (url) => ({
    timeout: {
        request: 10000,
    },
    agent: {
        https: ProxyAgent((0, proxy_from_env_1.getProxyForUrl)(url)),
    },
    url,
});
const getCurrentApiVersion = async () => (await (0, got_1.default)(getProxiedOptions('https://mdcoverage.secure.force.com/services/apexrest/report')).json()).versions.selected;
exports.getCurrentApiVersion = getCurrentApiVersion;
const getCoverage = async (apiVersion) => {
    const results = await Promise.allSettled(
    // one of these will match the current version, but they differ during the release cycle
    [44, 45, 46].map(async (na) => await (0, got_1.default)(getProxiedOptions(`https://na${na}.test1.pc-rnd.salesforce.com/mdcoverage/api.jsp`)).json()));
    for (const result of results) {
        if (result.status === 'fulfilled' && result.value?.apiVersion === apiVersion) {
            return result.value;
        }
    }
    throw new Error(`could not find coverage for api version ${apiVersion}`);
};
exports.getCoverage = getCoverage;
