// @salesforce/plugin-source/lib/commands/force/source/ignored/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceIgnoredCommand = void 0;
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const path = require("path");
const fs = require("fs");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-source', 'ignored_list');
class SourceIgnoredCommand extends command_1.SfdxCommand {
    /**
     * Outputs all forceignored files from package directories of a project,
     * or based on a sourcepath param that points to a specific file or directory.
     */
    // eslint-disable-next-line @typescript-eslint/require-await
    async run() {
        try {
            this.forceIgnore = source_deploy_retrieve_1.ForceIgnore.findAndCreate(this.project.getPath());
            const sourcepaths = this.flags.sourcepath
                ? [this.flags.sourcepath]
                : this.project.getUniquePackageDirectories().map((pDir) => pDir.path);
            const ignoredFiles = (await Promise.all(sourcepaths.map((sp) => this.statIgnored(sp.trim())))).flat();
            // Command output
            if (ignoredFiles.length) {
                this.ux.log('Found the following ignored files:');
                ignoredFiles.forEach((filepath) => this.ux.log(filepath));
            }
            else {
                this.ux.log('No ignored files found in paths:');
                sourcepaths.forEach((sp) => this.ux.log(sp));
            }
            return { ignoredFiles };
        }
        catch (err) {
            const error = err;
            if (error.code === 'ENOENT') {
                throw new core_1.SfError(messages.getMessage('invalidSourcePath', [this.flags.sourcepath]), 'invalidSourcePath');
            }
            throw core_1.SfError.wrap(error);
        }
    }
    // Stat the filepath.  Test if a file, recurse if a directory.
    async statIgnored(filepath) {
        const stats = await fs.promises.stat(filepath);
        if (stats.isDirectory()) {
            return (await Promise.all(await this.findIgnored(filepath))).flat();
        }
        else {
            return this.isIgnored(filepath) ? [filepath] : [];
        }
    }
    // Recursively search a directory for source files to test.
    async findIgnored(dir) {
        this.logger.debug(`Searching dir: ${dir}`);
        return (await fs.promises.readdir(dir)).map((filename) => this.statIgnored(path.join(dir, filename)));
    }
    // Test if a source file is denied, adding any ignored files to
    // the ignoredFiles array for output.
    isIgnored(filepath) {
        if (this.forceIgnore.denies(filepath)) {
            this.logger.debug(`[DENIED]: ${filepath}`);
            return true;
        }
        this.logger.debug(`[ACCEPTED]: ${filepath}`);
        return false;
    }
}
exports.SourceIgnoredCommand = SourceIgnoredCommand;
SourceIgnoredCommand.description = messages.getMessage('description');
SourceIgnoredCommand.requiresProject = true;
SourceIgnoredCommand.flagsConfig = {
    sourcepath: command_1.flags.filepath({
        char: 'p',
        description: messages.getMessage('flags.sourcepath'),
    }),
};
