// @salesforce/kit/lib/duration.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.sleep = exports.Duration = void 0;
/**
 * A simple utility class for converting durations between minutes, seconds, and milliseconds.
 */
class Duration {
    constructor(quantity, unit = Duration.Unit.MINUTES) {
        this.quantity = quantity;
        this.unit = unit;
    }
    /**
     * Returns the current number of minutes represented by this `Duration` instance, rounded to the nearest integer
     * value.
     */
    get minutes() {
        switch (this.unit) {
            case Duration.Unit.MILLISECONDS:
                return Math.round(this.quantity / Duration.MILLIS_IN_SECONDS / Duration.SECONDS_IN_MINUTE);
            case Duration.Unit.SECONDS:
                return Math.round(this.quantity / Duration.SECONDS_IN_MINUTE);
            case Duration.Unit.MINUTES:
                return this.quantity;
            case Duration.Unit.HOURS:
                return this.quantity * Duration.MINUTES_IN_HOUR;
            case Duration.Unit.DAYS:
                return this.quantity * Duration.MINUTES_IN_HOUR * Duration.HOURS_IN_DAY;
            case Duration.Unit.WEEKS:
                return this.quantity * Duration.MINUTES_IN_HOUR * Duration.HOURS_IN_DAY * Duration.DAYS_IN_WEEK;
        }
    }
    /**
     * Returns the current number of hours represented by this `Duration` instance.
     */
    get hours() {
        switch (this.unit) {
            case Duration.Unit.MILLISECONDS:
                return Math.round(this.quantity / Duration.MILLIS_IN_SECONDS / Duration.SECONDS_IN_MINUTE / Duration.MINUTES_IN_HOUR);
            case Duration.Unit.SECONDS:
                return Math.round(this.quantity / Duration.SECONDS_IN_MINUTE / Duration.MINUTES_IN_HOUR);
            case Duration.Unit.MINUTES:
                return Math.round(this.quantity / Duration.MINUTES_IN_HOUR);
            case Duration.Unit.HOURS:
                return this.quantity;
            case Duration.Unit.DAYS:
                return this.quantity * Duration.HOURS_IN_DAY;
            case Duration.Unit.WEEKS:
                return this.quantity * Duration.HOURS_IN_DAY * Duration.DAYS_IN_WEEK;
        }
    }
    /**
     * Returns the current number of milliseconds represented by this `Duration` instance.
     */
    get milliseconds() {
        switch (this.unit) {
            case Duration.Unit.MILLISECONDS:
                return this.quantity;
            case Duration.Unit.SECONDS:
                return this.quantity * Duration.MILLIS_IN_SECONDS;
            case Duration.Unit.MINUTES:
                return this.quantity * Duration.MILLIS_IN_SECONDS * Duration.SECONDS_IN_MINUTE;
            case Duration.Unit.HOURS:
                return this.quantity * Duration.MILLIS_IN_SECONDS * Duration.SECONDS_IN_MINUTE * Duration.MINUTES_IN_HOUR;
            case Duration.Unit.DAYS:
                return (this.quantity *
                    Duration.MILLIS_IN_SECONDS *
                    Duration.SECONDS_IN_MINUTE *
                    Duration.MINUTES_IN_HOUR *
                    Duration.HOURS_IN_DAY);
            case Duration.Unit.WEEKS:
                return (this.quantity *
                    Duration.MILLIS_IN_SECONDS *
                    Duration.SECONDS_IN_MINUTE *
                    Duration.MINUTES_IN_HOUR *
                    Duration.HOURS_IN_DAY *
                    Duration.DAYS_IN_WEEK);
        }
    }
    /**
     * Returns the current number of seconds represented by this `Duration` instance, rounded to the nearest integer
     * value.
     */
    get seconds() {
        switch (this.unit) {
            case Duration.Unit.MILLISECONDS:
                return Math.round(this.quantity / Duration.MILLIS_IN_SECONDS);
            case Duration.Unit.SECONDS:
                return this.quantity;
            case Duration.Unit.MINUTES:
                return this.quantity * Duration.SECONDS_IN_MINUTE;
            case Duration.Unit.HOURS:
                return this.quantity * Duration.SECONDS_IN_MINUTE * Duration.MINUTES_IN_HOUR;
            case Duration.Unit.DAYS:
                return this.quantity * Duration.SECONDS_IN_MINUTE * Duration.MINUTES_IN_HOUR * Duration.HOURS_IN_DAY;
            case Duration.Unit.WEEKS:
                return (this.quantity *
                    Duration.SECONDS_IN_MINUTE *
                    Duration.MINUTES_IN_HOUR *
                    Duration.HOURS_IN_DAY *
                    Duration.DAYS_IN_WEEK);
        }
    }
    /**
     * Returns the current number of days represented by this `Duration` instance.
     */
    get days() {
        switch (this.unit) {
            case Duration.Unit.MILLISECONDS:
                return Math.round(this.quantity /
                    Duration.MILLIS_IN_SECONDS /
                    Duration.SECONDS_IN_MINUTE /
                    Duration.MINUTES_IN_HOUR /
                    Duration.HOURS_IN_DAY);
            case Duration.Unit.SECONDS:
                return Math.round(this.quantity / Duration.SECONDS_IN_MINUTE / Duration.MINUTES_IN_HOUR / Duration.HOURS_IN_DAY);
            case Duration.Unit.MINUTES:
                return Math.round(this.quantity / Duration.MINUTES_IN_HOUR / Duration.HOURS_IN_DAY);
            case Duration.Unit.HOURS:
                return Math.round(this.quantity / Duration.HOURS_IN_DAY);
            case Duration.Unit.DAYS:
                return this.quantity;
            case Duration.Unit.WEEKS:
                return this.quantity * Duration.DAYS_IN_WEEK;
        }
    }
    /**
     * Returns the current number of weeks represented by this `Duration` instance.
     */
    get weeks() {
        switch (this.unit) {
            case Duration.Unit.MILLISECONDS:
                return Math.round(this.quantity /
                    Duration.MILLIS_IN_SECONDS /
                    Duration.SECONDS_IN_MINUTE /
                    Duration.MINUTES_IN_HOUR /
                    Duration.HOURS_IN_DAY /
                    Duration.DAYS_IN_WEEK);
            case Duration.Unit.SECONDS:
                return Math.round(this.quantity /
                    Duration.SECONDS_IN_MINUTE /
                    Duration.MINUTES_IN_HOUR /
                    Duration.HOURS_IN_DAY /
                    Duration.DAYS_IN_WEEK);
            case Duration.Unit.MINUTES:
                return Math.round(this.quantity / Duration.MINUTES_IN_HOUR / Duration.HOURS_IN_DAY / Duration.DAYS_IN_WEEK);
            case Duration.Unit.HOURS:
                return Math.round(this.quantity / Duration.HOURS_IN_DAY / Duration.DAYS_IN_WEEK);
            case Duration.Unit.DAYS:
                return Math.round(this.quantity / Duration.DAYS_IN_WEEK);
            case Duration.Unit.WEEKS:
                return this.quantity;
        }
    }
    /**
     * Returns a new `Duration` instance created from the specified number of milliseconds.
     *
     * @param quantity The number of milliseconds.
     */
    static milliseconds(quantity) {
        return new Duration(quantity, Duration.Unit.MILLISECONDS);
    }
    /**
     * Returns a new `Duration` instance created from the specified number of seconds.
     *
     * @param quantity The number of seconds.
     */
    static seconds(quantity) {
        return new Duration(quantity, Duration.Unit.SECONDS);
    }
    /**
     * Returns a new `Duration` instance created from the specified number of minutes.
     *
     * @param quantity The number of minutes.
     */
    static minutes(quantity) {
        return new Duration(quantity, Duration.Unit.MINUTES);
    }
    /**
     * Returns a new `Duration` instance created from the specified number of hours.
     *
     * @param quantity The number of hours.
     */
    static hours(quantity) {
        return new Duration(quantity, Duration.Unit.HOURS);
    }
    /**
     * Returns a new `Duration` instance created from the specified number of days.
     *
     * @param quantity The number of days.
     */
    static days(quantity) {
        return new Duration(quantity, Duration.Unit.DAYS);
    }
    /**
     * Returns a new `Duration` instance created from the specified number of weeks.
     *
     * @param quantity The number of weeks.
     */
    static weeks(quantity) {
        return new Duration(quantity, Duration.Unit.WEEKS);
    }
    /**
     * The string representation of this `Duration`. e.g. "645 seconds"
     */
    toString() {
        return this.pluralize();
    }
    pluralize(num = this.quantity, unit = this.unit) {
        const name = Duration.Unit[unit].toLowerCase();
        if (num === 1)
            return `${num} ${name.slice(0, name.length - 1)}`;
        return `${num} ${name}`;
    }
}
exports.Duration = Duration;
/**
 * The number of milliseconds in one second.
 */
Duration.MILLIS_IN_SECONDS = 1000;
/**
 * The number of seconds in one minute.
 */
Duration.SECONDS_IN_MINUTE = 60;
/**
 * The number of minutes in one hour.
 */
Duration.MINUTES_IN_HOUR = 60;
/**
 * The number of hours in one day.
 */
Duration.HOURS_IN_DAY = 24;
/**
 * The number of days in one week.
 */
Duration.DAYS_IN_WEEK = 7;
(function (Duration) {
    /**
     * Units of duration.
     */
    // eslint-disable-next-line no-shadow
    let Unit;
    (function (Unit) {
        Unit[Unit["MINUTES"] = 0] = "MINUTES";
        Unit[Unit["MILLISECONDS"] = 1] = "MILLISECONDS";
        Unit[Unit["SECONDS"] = 2] = "SECONDS";
        Unit[Unit["HOURS"] = 3] = "HOURS";
        Unit[Unit["DAYS"] = 4] = "DAYS";
        Unit[Unit["WEEKS"] = 5] = "WEEKS";
    })(Unit = Duration.Unit || (Duration.Unit = {}));
})(Duration = exports.Duration || (exports.Duration = {}));
// underlying function
function sleep(durationOrQuantity, unit = Duration.Unit.MILLISECONDS) {
    const duration = durationOrQuantity instanceof Duration ? durationOrQuantity : new Duration(durationOrQuantity, unit);
    let handle;
    let doResolve;
    const wake = () => {
        if (!handle)
            return;
        clearTimeout(handle);
        handle = undefined;
        doResolve();
    };
    const promise = new Promise((resolve) => {
        doResolve = resolve;
        handle = setTimeout(wake, duration.milliseconds);
    });
    return Object.assign(promise, { interrupt: wake });
}
exports.sleep = sleep;
