// @salesforce/plugin-signups/lib/commands/force/org/snapshot/get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotGet = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const os_1 = require("os");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const snapshot_1 = require("../../../../shared/snapshot");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-signups', 'snapshot.get');
class SnapshotGet extends command_1.SfdxCommand {
    async run() {
        const result = await (0, snapshot_1.queryByNameOrId)(this.hubOrg.getConnection(), this.flags.snapshot);
        if (!this.flags.json) {
            (0, snapshot_1.printSingleRecordTable)(result);
        }
        return result;
    }
}
exports.SnapshotGet = SnapshotGet;
SnapshotGet.description = messages.getMessage('description');
SnapshotGet.examples = messages.getMessage('examples').split(os_1.EOL);
SnapshotGet.requiresDevhubUsername = true;
SnapshotGet.flagsConfig = {
    snapshot: command_1.flags.string({
        char: 's',
        description: messages.getMessage('flags.snapshot'),
        longDescription: messages.getMessage('flagsLong.snapshot'),
        required: true,
    }),
};
