// @salesforce/apex-node/lib/src/logs/index.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LISTENER_ABORTED_ERROR_NAME = exports.LOG_TIMER_LENGTH_MINUTES = exports.MAX_NUM_LOGS = exports.LOG_TYPE = exports.DEFAULT_DEBUG_LEVEL_NAME = exports.LogService = void 0;
var logService_1 = require("./logService");
Object.defineProperty(exports, "LogService", { enumerable: true, get: function () { return logService_1.LogService; } });
var constants_1 = require("./constants");
Object.defineProperty(exports, "DEFAULT_DEBUG_LEVEL_NAME", { enumerable: true, get: function () { return constants_1.DEFAULT_DEBUG_LEVEL_NAME; } });
Object.defineProperty(exports, "LOG_TYPE", { enumerable: true, get: function () { return constants_1.LOG_TYPE; } });
Object.defineProperty(exports, "MAX_NUM_LOGS", { enumerable: true, get: function () { return constants_1.MAX_NUM_LOGS; } });
Object.defineProperty(exports, "LOG_TIMER_LENGTH_MINUTES", { enumerable: true, get: function () { return constants_1.LOG_TIMER_LENGTH_MINUTES; } });
Object.defineProperty(exports, "LISTENER_ABORTED_ERROR_NAME", { enumerable: true, get: function () { return constants_1.LISTENER_ABORTED_ERROR_NAME; } });
