// @salesforce/plugin-templates/lib/commands/force/lightning/test/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const command_1 = require("@salesforce/command");
const lightningTestGenerator_1 = require("@salesforce/templates/lib/generators/lightningTestGenerator");
const utils_1 = require("@salesforce/templates/lib/utils");
const utils_2 = require("../../../../utils");
const lightningTestFileSuffix = /.resource$/;
class LightningTest extends utils_2.TemplateCommand {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.runGenerator(lightningTestGenerator_1.default);
        });
    }
}
exports.default = LightningTest;
LightningTest.description = utils_2.MessageUtil.buildDescription('LightningTestDescription', false);
LightningTest.examples = [
    '$ sfdx force:lightning:test:create -n MyLightningTest',
    '$ sfdx force:lightning:test:create -n MyLightningTest -d lightningTests'
];
LightningTest.help = utils_2.MessageUtil.buildHelpText(LightningTest.examples, false);
LightningTest.longDescription = utils_2.MessageUtil.get('LightningTestLongDescription');
LightningTest.flagsConfig = {
    testname: command_1.flags.string({
        char: 'n',
        description: utils_2.MessageUtil.get('LightningNameFlagDescription', [
            utils_2.MessageUtil.get('Test')
        ]),
        longDescription: utils_2.MessageUtil.get('LightningTestNameFlagLongDescription'),
        required: true
    }),
    template: command_1.flags.string({
        char: 't',
        description: utils_2.MessageUtil.get('TemplateFlagDescription'),
        longDescription: utils_2.MessageUtil.get('TemplateFlagLongDescription'),
        default: 'DefaultLightningTest',
        options: utils_1.CreateUtil.getCommandTemplatesForFiletype(lightningTestFileSuffix, 'lightningtest')
    }),
    outputdir: command_1.flags.string({
        char: 'd',
        description: utils_2.MessageUtil.get('OutputDirFlagDescription'),
        longDescription: utils_2.MessageUtil.get('OutputDirFlagLongDescription'),
        default: '.'
    }),
    internal: command_1.flags.boolean({
        char: 'i',
        description: utils_2.MessageUtil.get('LightningInternalFlagDescription'),
        hidden: true
    })
};
