// @salesforce/lazy-require/lib/LazyRequire.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const LazyLoader_1 = require("./LazyLoader");
const TypeCache_1 = require("./TypeCache");
class LazyRequire {
    constructor(typeCache, exclusions) {
        this.typeCache = typeCache;
        this.lazyLoader = new LazyLoader_1.default(this.typeCache, exclusions);
    }
    static create(typeCacheFile) {
        const typeCache = new TypeCache_1.default(fs, typeCacheFile);
        return new LazyRequire(typeCache);
    }
    /**
     * Resets the type cache, whether or not lazy loading has been started.
     */
    resetTypeCache() {
        return this.typeCache.reset();
    }
    /**
     * Start lazy loading type-compatible modules.
     */
    start() {
        if (this.lazyLoader.isEnabled()) {
            return;
        }
        this.typeCache.load();
        process.on('exit', () => this.typeCache.save());
        this.lazyLoader.enable();
    }
}
exports.default = LazyRequire;
