// jsforce/browser/browser/jsonp.js
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _setTimeout from "@babel/runtime-corejs3/core-js-stable/set-timeout";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _indexOfInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/index-of";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
import "regenerator-runtime/runtime";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";

/**
 *
 */
import { Transform } from 'stream';
var _index = 0;

function processJsonpRequest(_x, _x2, _x3) {
  return _processJsonpRequest.apply(this, arguments);
}

function _processJsonpRequest() {
  _processJsonpRequest = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(params, jsonpParam, timeout) {
    var _context2;

    var cbFuncName, callbacks, url, script, pid, res;
    return _regeneratorRuntime.wrap(function _callee2$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            if (!(params.method.toUpperCase() !== 'GET')) {
              _context3.next = 2;
              break;
            }

            throw new Error('JSONP only supports GET request.');

          case 2:
            _index += 1;
            cbFuncName = "_jsforce_jsonpCallback_".concat(_index);
            callbacks = window;
            url = params.url;
            url += _indexOfInstanceProperty(url).call(url, '?') > 0 ? '&' : '?';
            url += _concatInstanceProperty(_context2 = "".concat(jsonpParam, "=")).call(_context2, cbFuncName);
            script = document.createElement('script');
            script.type = 'text/javascript';
            script.src = url;

            if (document.documentElement) {
              document.documentElement.appendChild(script);
            }

            _context3.prev = 12;
            _context3.next = 15;
            return new _Promise(function (resolve, reject) {
              pid = _setTimeout(function () {
                reject(new Error('JSONP call time out.'));
              }, timeout);
              callbacks[cbFuncName] = resolve;
            });

          case 15:
            res = _context3.sent;
            return _context3.abrupt("return", {
              statusCode: 200,
              headers: {
                'content-type': 'application/json'
              },
              body: _JSON$stringify(res)
            });

          case 17:
            _context3.prev = 17;
            clearTimeout(pid);

            if (document.documentElement) {
              document.documentElement.removeChild(script);
            }

            delete callbacks[cbFuncName];
            return _context3.finish(17);

          case 22:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee2, null, [[12,, 17, 22]]);
  }));
  return _processJsonpRequest.apply(this, arguments);
}

function createRequest() {
  var jsonpParam = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'callback';
  var timeout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 10000;
  return function (params) {
    var stream = new Transform({
      transform: function transform(chunk, encoding, callback) {
        callback();
      },
      flush: function flush() {
        _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
          var response;
          return _regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  _context.next = 2;
                  return processJsonpRequest(params, jsonpParam, timeout);

                case 2:
                  response = _context.sent;
                  stream.emit('response', response);
                  stream.emit('complete', response);
                  stream.push(response.body);
                  stream.push(null);

                case 7:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee);
        }))();
      }
    });
    stream.end();
    return stream;
  };
}

export default {
  supported: typeof window !== 'undefined' && typeof document !== 'undefined',
  createRequest: createRequest
};
